<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial and basic overview of the admixr R package • admixr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial and basic overview of the admixr R package">
<meta property="og:description" content="admixr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">admixr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.9.1.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01-tutorial.html">Tutorial and basic overview of the admixr R package</a>
    </li>
    <li>
      <a href="../articles/02-qpAdm.html">Fitting qpAdm models with a 'rotation' strategy</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bodkan/admixr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tutorial and basic overview of the admixr R
package</h1>
                        <h4 data-toc-skip class="author">Martin
Petr</h4>
            
            <h4 data-toc-skip class="date">2024-05-08</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/admixr/blob/HEAD/vignettes/01-tutorial.Rmd" class="external-link"><code>vignettes/01-tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>01-tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><a href="https://github.com/DReichLab/AdmixTools/" class="external-link">ADMIXTOOLS</a> is
a widely used software package for calculating admixture statistics and
testing population admixture hypotheses.</p>
<p>A typical ADMIXTOOLS workflow generally involves a combination of
<code>sed</code>/<code>awk</code>/shell scripting and manual editing to
create text configuration files. These are then passed as command-line
arguments to one of ADMIXTOOLS commands, and control how to run a
particular analysis. The results are then redirected to another file,
which has to be parsed by the user to extract values of interest, often
using command-line utilities again or (worse) by manual copy-pasting.
Finally, the processed results are analysed in R, Excel or another
program.</p>
<p>This workflow can be a little cumbersome, especially if one wants to
explore many hypotheses involving different combinations of populations.
Most importantly, however, it makes it difficult to coduct reproducible
research, as it is nearly impossible to construct fully automated
“pipelines” that don’t require user intervention.</p>
<p>This R package makes it possible to perform all stages of ADMIXTOOLS
analyses entirely from R, completely removing the need for “low level”
configuration of individual ADMIXTOOLS programs.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><strong>Note that in order to use the <em>admixr</em> package, you
need a working installation of ADMIXTOOLS!</strong> You can find
installation instructions <a href="https://github.com/DReichLab/AdmixTools/blob/master/README.INSTALL" class="external-link">here</a>.
The software runs on Linux and macOS and these are the two systems that
<em>admixr</em> is tested on.</p>
<p><strong>Furthermore, you also need to make sure that R can find
ADMIXTOOLS binaries on the <code>$PATH</code>.</strong> You can achieve
this by specifying
<code>PATH=&lt;path to the location of ADMIXTOOLS programs&gt;</code> in
the <code>.Renviron</code> file.</p>
<p>To install <em>admixr</em> you can simply run the following command
in your R session:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"admixr"</span><span class="op">)</span></span></code></pre></div>
<p>Furthermore, if you want to follow the examples in this vignette, you
will need the <a href="https://www.tidyverse.org" class="external-link">tidyverse</a>
collection of packages for data manipulation manipulation and plotting,
which you can install with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span></span></code></pre></div>
<p>You definitely don’t need tidyverse for working with <em>admixr</em>
but it really makes data manipulation and plotting things much easier. I
recommend at least giving it a shot.</p>
<p>When everything is ready, you can run the following code to load both
packages:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/admixr" class="external-link">admixr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="a-note-about-eigenstrat-format">A note about EIGENSTRAT format<a class="anchor" aria-label="anchor" href="#a-note-about-eigenstrat-format"></a>
</h2>
<p>ADMIXTOOLS software uses a peculiar set of genetic file formats,
which may seem strange if you are used to working with <a href="http://samtools.github.io/hts-specs/VCFv4.3.pdf" class="external-link">VCF files</a>.
However, the basic idea remains the same: we want to store and access
SNP data (REF/ALT alleles) of a set of individuals at a defined set of
genomic positions.</p>
<p>EIGENSTRAT datasets always contain three kinds of files:</p>
<ul>
<li>
<code>ind</code> file - specifies a unique name, sex (optional - can
be simply “U” for “undefined”) and label (such as population assignment)
of each sample;</li>
<li>
<code>snp</code> file - specifies the positions of SNPs, REF/ALT
alleles etc.;</li>
<li>
<code>geno</code> file - contains SNP data (one row per site, one
character per sample) in a dense string-based format:
<ul>
<li>0: individual is homozygous ALT</li>
<li>1: individual is a heterozygote</li>
<li>2: individual is homozygous REF</li>
<li>9: missing data</li>
</ul>
</li>
</ul>
<p>Therefore, a VCF file is essentially a combination of all three files
in a single package.</p>
<p>Let’s first download a small testing SNP dataset using a built-in
<em>admixr</em> function <code><a href="../reference/download_data.html">download_data()</a></code>. This function
downloads the data into a temporary directory (you can specify the
destination using its <code>dirname</code> argument, in case you want to
place it elsewhere). In addition to this, the function returns a shared
path/prefix of the whole dataset.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>(prefix <span class="ot">&lt;-</span> <span class="fu">download_data</span>(<span class="at">dirname =</span> <span class="fu">tempdir</span>()))</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"/var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps"</span></span></code></pre></div>
<p>We can verify that there are indeed three files with this prefix:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">list.files</span>(<span class="at">path =</span> <span class="fu">dirname</span>(prefix), <span class="at">pattern =</span> <span class="fu">basename</span>(prefix), <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"/var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.geno"</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>[<span class="dv">2</span>] <span class="st">"/var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.ind"</span> </span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>[<span class="dv">3</span>] <span class="st">"/var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.snp"</span> </span></code></pre></div>
<p>Let’s look at their contents:</p>
<div class="section level4">
<h4 id="ind-file">
<code>ind</code> file<a class="anchor" aria-label="anchor" href="#ind-file"></a>
</h4>
<pre><code>Chimp        U  Chimp
Mbuti        U  Mbuti
Yoruba       U  Yoruba
Khomani_San  U  Khomani_San
Han          U  Han
Dinka        U  Dinka
Sardinian    U  Sardinian
Papuan       U  Papuan
French       U  French
Vindija      U  Vindija
Altai        U  Altai
Denisova     U  Denisova</code></pre>
<p>The first column (sample name) and the third column (population
label) are generally not the same (sample names often have numerical
suffixes to make them unique, etc.), but were kept the same here for
simplicity. Importantly, when specifying population/sample names in
<em>admixr</em> functions, the information in the third column is what
is used. For example, if you have individuals such as “French1”,
“French2”, “French3” in the first column of an <code>ind</code> file,
all three sharing a “French” population label in the third column,
specifying “French” in an <em>admixr</em> function will combine all
three samples in a single population, instead of working with each
individual separately.</p>
</div>
<div class="section level4">
<h4 id="snp-file-first-3-lines">
<code>snp</code> file (first 3 lines)<a class="anchor" aria-label="anchor" href="#snp-file-first-3-lines"></a>
</h4>
<pre><code>1_832756    1   0.008328    832756  T   G
1_838931    1   0.008389    838931  A   C
1_843249    1   0.008432    843249  A   T</code></pre>
<p>The columns of this file are, in order:</p>
<ol style="list-style-type: decimal">
<li>SNP string ID</li>
<li>chromosome</li>
<li>genetic distance</li>
<li>position along a chromosome</li>
<li>reference allele</li>
<li>alternative allele</li>
</ol>
</div>
<div class="section level4">
<h4 id="geno-file-first-3-lines">
<code>geno</code> file (first 3 lines)<a class="anchor" aria-label="anchor" href="#geno-file-first-3-lines"></a>
</h4>
<pre><code><span><span class="fl">902021012000</span></span>
<span><span class="fl">922221211222</span></span>
<span><span class="fl">922222122222</span></span></code></pre>
<p>Each row is one genomic site, each column is a genotype in one
individual.</p>
</div>
</div>
<div class="section level2">
<h2 id="philosophy-of-admixr">Philosophy of <em>admixr</em><a class="anchor" aria-label="anchor" href="#philosophy-of-admixr"></a>
</h2>
<p>The goal of <em>admixr</em> is to make ADMIXTOOLS analyses as trivial
to run as possible, without having to worry about par/pop/left/right
configuration files (as they are known in the jargon of ADMIXTOOLS) and
other low-level details.</p>
<p>The only interface between you and ADMIXTOOLS is the following set of
R functions:</p>
<ul>
<li><code><a href="../reference/f4ratio.html">d()</a></code></li>
<li><code><a href="../reference/f4ratio.html">f4()</a></code></li>
<li><code><a href="../reference/f4ratio.html">f4ratio()</a></code></li>
<li><code><a href="../reference/f4ratio.html">f3()</a></code></li>
<li><code><a href="../reference/qpAdm.html">qpAdm()</a></code></li>
<li><code><a href="../reference/qpWave.html">qpWave()</a></code></li>
</ul>
<p>Anything that would normally require <a href="https://gaworkshop.readthedocs.io/en/latest/contents/06_f3/f3.html" class="external-link">dozens
of lines of shell scripts</a> can be often accomplished by running a
single line of R code.</p>
</div>
<div class="section level2">
<h2 id="internal-representation-of-eigenstrat-data">Internal representation of EIGENSTRAT data<a class="anchor" aria-label="anchor" href="#internal-representation-of-eigenstrat-data"></a>
</h2>
<p>As we saw above, each EIGENSTRAT dataset has three components. The
way this data is internally represented in <em>admixr</em> is using a
small S3 R object created using the <code>eigenstrat</code> constructor
function. This function accepts the path and prefix of a trio of
EIGENSTRAT snp/ind/geno files and returns an R object of the class
<code>EIGENSTRAT</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/eigenstrat.html">eigenstrat</a></span><span class="op">(</span><span class="va">prefix</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps</span></span>
<span><span class="co">#&gt; EIGENSTRAT object</span></span>
<span><span class="co">#&gt; =================</span></span>
<span><span class="co">#&gt; components:</span></span>
<span><span class="co">#&gt;   ind file: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.ind</span></span>
<span><span class="co">#&gt;   snp file: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.snp</span></span>
<span><span class="co">#&gt;   geno file: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.geno</span></span></code></pre></div>
<p>This object encapsulates the paths to all three EIGENSTRAT components
and makes it easy to pass the data to different <em>admixr</em>
functions.</p>
<p>The following couple of sections describe how to use the
<em>admixr</em> package using simple example analyses.</p>
</div>
<div class="section level2">
<h2 id="d-statistic">D statistic<a class="anchor" aria-label="anchor" href="#d-statistic"></a>
</h2>
<p>Let’s say we are interested in the following question: <em>“Which
populations today show evidence of Neanderthal admixture?</em></p>
<p>One way of looking at this is using the following D statistic: <span class="math display">\[D(\textrm{present-day human W}, \textrm{African},
\textrm{Neanderthal}, \textrm{Chimp}).\]</span></p>
<p><span class="math inline">\(D\)</span> statistics are based on
comparing the proportions of BABA and ABBA sites patterns observed in
the data:</p>
<p><span class="math display">\[D = \frac{\textrm{# BABA sites - # ABBA
sites}}{\textrm{# BABA sites + # ABBA sites}}.\]</span></p>
<p>Significant departure of <span class="math inline">\(D\)</span> from
zero indicates an excess of allele sharing between the first and the
third population (positive <span class="math inline">\(D\)</span>), or
an excess of allele sharing between the second and the third population
(negative <span class="math inline">\(D\)</span>). If we get <span class="math inline">\(D\)</span> that is not significantly different
from 0, this suggests that the first and second populations form a
clade, and don’t differ in the rate of allele sharing with the third
population (this is the null hypothesis that the data is compared
against).</p>
<p>Therefore, our <span class="math inline">\(D\)</span> statistic above
tests whether some modern humans today admixed with Neanderthals, which
would increase their genetic affinity to this archaic group compared to
Africans (whose ancestors never met Neanderthals).</p>
<p>Let’s save some population names first to make our code more
concise:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pops</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"French"</span>, <span class="st">"Sardinian"</span>, <span class="st">"Han"</span>, <span class="st">"Papuan"</span>, <span class="st">"Khomani_San"</span>, <span class="st">"Mbuti"</span>, <span class="st">"Dinka"</span><span class="op">)</span></span></code></pre></div>
<p>Using the <em>admixr</em> package we can then calculate our <span class="math inline">\(D\)</span> statistic simply by running:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/f4ratio.html">d</a></span><span class="op">(</span>W <span class="op">=</span> <span class="va">pops</span>, X <span class="op">=</span> <span class="st">"Yoruba"</span>, Y <span class="op">=</span> <span class="st">"Vindija"</span>, Z <span class="op">=</span> <span class="st">"Chimp"</span>, data <span class="op">=</span> <span class="va">snps</span><span class="op">)</span></span></code></pre></div>
<p>The result is a following data frame:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="15%">
<col width="9%">
<col width="10%">
<col width="7%">
<col width="10%">
<col width="11%">
<col width="9%">
<col width="7%">
<col width="7%">
<col width="9%">
</colgroup>
<thead><tr class="header">
<th align="left">W</th>
<th align="left">X</th>
<th align="left">Y</th>
<th align="left">Z</th>
<th align="right">D</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
<th align="right">BABA</th>
<th align="right">ABBA</th>
<th align="right">nsnps</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">French</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0313</td>
<td align="right">0.006933</td>
<td align="right">4.510</td>
<td align="right">15802</td>
<td align="right">14844</td>
<td align="right">487753</td>
</tr>
<tr class="even">
<td align="left">Sardinian</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0287</td>
<td align="right">0.006792</td>
<td align="right">4.222</td>
<td align="right">15729</td>
<td align="right">14852</td>
<td align="right">487646</td>
</tr>
<tr class="odd">
<td align="left">Han</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0278</td>
<td align="right">0.006609</td>
<td align="right">4.199</td>
<td align="right">15780</td>
<td align="right">14928</td>
<td align="right">487925</td>
</tr>
<tr class="even">
<td align="left">Papuan</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0457</td>
<td align="right">0.006571</td>
<td align="right">6.953</td>
<td align="right">16131</td>
<td align="right">14721</td>
<td align="right">487694</td>
</tr>
<tr class="odd">
<td align="left">Khomani_San</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0066</td>
<td align="right">0.006292</td>
<td align="right">1.051</td>
<td align="right">16168</td>
<td align="right">15955</td>
<td align="right">487564</td>
</tr>
<tr class="even">
<td align="left">Mbuti</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">-0.0005</td>
<td align="right">0.006345</td>
<td align="right">-0.074</td>
<td align="right">15751</td>
<td align="right">15766</td>
<td align="right">487642</td>
</tr>
</tbody>
</table>
<p>We can see that in addition to the specified population names, the
output table contains additional columns:</p>
<ul>
<li>
<code>D</code> - <span class="math inline">\(D\)</span> statistic
value</li>
<li>
<code>stderr</code> - standard error of the <span class="math inline">\(D\)</span> statistic calculated using the block
jackknife</li>
<li>
<code>Zscore</code> - <span class="math inline">\(Z\)</span>-zscore
value (number of standard errors the <span class="math inline">\(D\)</span> is from 0, i.e. how strongly do we
reject the null hypothesis of no admixture)</li>
<li>
<code>BABA</code>, <code>ABBA</code> - counts of observed site
patterns</li>
<li>
<code>nsnps</code> - number of SNPs used for a given
calculation</li>
</ul>
<p>While we could certainly make inferences by looking at the <span class="math inline">\(Z\)</span>-scores, tables in general are not the
best representation of this kind of data, especially as the number of
samples increases. Instead, we can use the <a href="https://ggplot2.tidyverse.org" class="external-link"><code>ggplot2</code></a> package to
plot the results:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">result</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_reorder.html" class="external-link">fct_reorder</a></span><span class="op">(</span><span class="va">W</span>, <span class="va">D</span><span class="op">)</span>, <span class="va">D</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">Zscore</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_errorbar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>ymin <span class="op">=</span> <span class="va">D</span> <span class="op">-</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">stderr</span>, ymax <span class="op">=</span> <span class="va">D</span> <span class="op">+</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">stderr</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="01-tutorial_files/figure-html/d_plot-1.png" width="672"></p>
<p>(If you want to more know about data analysis using R, including
plotting with ggplot2, I highly recommend <a href="https://r4ds.had.co.nz/" class="external-link">this</a> free book.)</p>
<p>We can see that the <span class="math inline">\(D\)</span> values for
Africans are not significantly different from 0, meaning that the data
is consistent with the null hypothesis of no Neanderthal ancestry in
Africans. On the other hand, the test rejects the null hypothesis for
all non-Africans today, suggesting that Neanderthals admixed with the
ancestors of present-day non-Africans.</p>
</div>
<div class="section level2">
<h2 id="f4-statistic">f4 statistic<a class="anchor" aria-label="anchor" href="#f4-statistic"></a>
</h2>
<p>An alternative way of addressing the previous question is to use the
<span class="math inline">\(f_4\)</span> statistic, which is very
similar to <span class="math inline">\(D\)</span> statistic and can be
calculated as:</p>
<p><span class="math display">\[ f_4 = \frac{\textrm{# BABA sites - #
ABBA sites}}{\textrm{# sites}}\]</span></p>
<p>Again, significant departure of <span class="math inline">\(f_4\)</span> from 0 can be interpreted as evidence
of gene flow.</p>
<p>To repeat the previous analysis using <span class="math inline">\(f_4\)</span> statistic, we can run the function
<code><a href="../reference/f4ratio.html">f4()</a></code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/f4ratio.html">f4</a></span><span class="op">(</span>W <span class="op">=</span> <span class="va">pops</span>, X <span class="op">=</span> <span class="st">"Yoruba"</span>, Y <span class="op">=</span> <span class="st">"Vindija"</span>, Z <span class="op">=</span> <span class="st">"Chimp"</span>, data <span class="op">=</span> <span class="va">snps</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="15%">
<col width="8%">
<col width="10%">
<col width="7%">
<col width="12%">
<col width="11%">
<col width="8%">
<col width="7%">
<col width="7%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="left">W</th>
<th align="left">X</th>
<th align="left">Y</th>
<th align="left">Z</th>
<th align="right">f4</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
<th align="right">BABA</th>
<th align="right">ABBA</th>
<th align="right">nsnps</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">French</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.001965</td>
<td align="right">0.000437</td>
<td align="right">4.501</td>
<td align="right">15802</td>
<td align="right">14844</td>
<td align="right">487753</td>
</tr>
<tr class="even">
<td align="left">Sardinian</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.001798</td>
<td align="right">0.000427</td>
<td align="right">4.209</td>
<td align="right">15729</td>
<td align="right">14852</td>
<td align="right">487646</td>
</tr>
<tr class="odd">
<td align="left">Han</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.001746</td>
<td align="right">0.000418</td>
<td align="right">4.178</td>
<td align="right">15780</td>
<td align="right">14928</td>
<td align="right">487925</td>
</tr>
<tr class="even">
<td align="left">Papuan</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.002890</td>
<td align="right">0.000417</td>
<td align="right">6.924</td>
<td align="right">16131</td>
<td align="right">14721</td>
<td align="right">487694</td>
</tr>
<tr class="odd">
<td align="left">Khomani_San</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.000436</td>
<td align="right">0.000415</td>
<td align="right">1.051</td>
<td align="right">16168</td>
<td align="right">15955</td>
<td align="right">487564</td>
</tr>
<tr class="even">
<td align="left">Mbuti</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">-0.000030</td>
<td align="right">0.000410</td>
<td align="right">-0.074</td>
<td align="right">15751</td>
<td align="right">15766</td>
<td align="right">487642</td>
</tr>
</tbody>
</table>
<p>By comparing this result to the <span class="math inline">\(D\)</span> statistic analysis above, we can make
the same conclusions.</p>
<p>You might be wondering why we have both <span class="math inline">\(f_4\)</span> and <span class="math inline">\(D\)</span> if they are so similar. The truth is
that <span class="math inline">\(f_4\)</span> is, among other things,
directly informative about the amount of shared genetic drift (“branch
length”) between pairs of populations, which is a very useful
theoretical property. Other than that, it’s often a matter of personal
preference and so <em>admixr</em> provides functions for calculating
both.</p>
</div>
<div class="section level2">
<h2 id="f4-ratio-statistic">f4-ratio statistic<a class="anchor" aria-label="anchor" href="#f4-ratio-statistic"></a>
</h2>
<p>Now we know that non-Africans today carry <em>some</em> Neanderthal
ancestry. But what if we want to know <em>how much</em> Neanderthal
ancestry they have? What proportion of their genomes is of Neanderthal
origin? To answer questions like this, we can use the <span class="math inline">\(f_4\)</span>-ratio statistic, which can be
formulated in the following way (using a notation of <a href="http://www.genetics.org/content/192/3/1065" class="external-link">Patterson et al.,
2012</a>, who formally described its properties).</p>
<p><span class="math display">\[f_4\textrm{-ratio} = \frac{f_4(A, O; X,
C)}{f_4(A, O; B, C)}.\]</span></p>
<p>Using <code>amidxr</code>, we can calculate <span class="math inline">\(f_4\)</span>-ratios using the following code
(<code>X</code> being a vector of samples which we want to estimate the
Neanderthal ancestry in):</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/f4ratio.html">f4ratio</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">pops</span>, A <span class="op">=</span> <span class="st">"Altai"</span>, B <span class="op">=</span> <span class="st">"Vindija"</span>, C <span class="op">=</span> <span class="st">"Yoruba"</span>, O <span class="op">=</span> <span class="st">"Chimp"</span>, data <span class="op">=</span> <span class="va">snps</span><span class="op">)</span></span></code></pre></div>
<p>The ancestry proportion (a number between 0 and 1) is given in the
<code>alpha</code> column:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="9%">
<col width="12%">
<col width="18%">
<col width="10%">
<col width="9%">
<col width="14%">
<col width="14%">
<col width="10%">
</colgroup>
<thead><tr class="header">
<th align="left">A</th>
<th align="left">B</th>
<th align="left">X</th>
<th align="left">C</th>
<th align="left">O</th>
<th align="right">alpha</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">French</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.023774</td>
<td align="right">0.006176</td>
<td align="right">3.850</td>
</tr>
<tr class="even">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Sardinian</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.024468</td>
<td align="right">0.006071</td>
<td align="right">4.031</td>
</tr>
<tr class="odd">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Han</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.022117</td>
<td align="right">0.005892</td>
<td align="right">3.754</td>
</tr>
<tr class="even">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Papuan</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.037311</td>
<td align="right">0.005812</td>
<td align="right">6.420</td>
</tr>
<tr class="odd">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Khomani_San</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.003909</td>
<td align="right">0.005913</td>
<td align="right">0.661</td>
</tr>
<tr class="even">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Mbuti</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.000319</td>
<td align="right">0.005717</td>
<td align="right">0.056</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">result</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_reorder.html" class="external-link">fct_reorder</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">alpha</span><span class="op">)</span>, <span class="va">alpha</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">Zscore</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_errorbar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>ymin <span class="op">=</span> <span class="va">alpha</span> <span class="op">-</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">stderr</span>, ymax <span class="op">=</span> <span class="va">alpha</span> <span class="op">+</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">stderr</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Neandertal ancestry proportion"</span>, x <span class="op">=</span> <span class="st">"present-day individual"</span><span class="op">)</span></span></code></pre></div>
<p><img src="01-tutorial_files/figure-html/f4ratio_plot-1.png" width="672"></p>
<p>We can make several observations:</p>
<ul>
<li>Again, we don’t see any significant Neanderthal ancestry in
present-day Africans (proportion is consistent with 0%), which is what
we confirmed using <span class="math inline">\(D\)</span> and <span class="math inline">\(f_4\)</span> above.</li>
<li>Present-day non-Africans carry between 2-3% of Neanderthal
ancestry.</li>
<li>We see a much higher proportion of Neanderthal ancestry in people
from Papua New Guinea - more than 4%. This is consistent with earlier
studies that suggest additional archaic admixture events in the
ancestors of present-day Papuans.</li>
</ul>
</div>
<div class="section level2">
<h2 id="f3-statistic">f3 statistic<a class="anchor" aria-label="anchor" href="#f3-statistic"></a>
</h2>
<p>The <span class="math inline">\(f_3\)</span> statistic, also known as
the 3-population statistic, is useful whenever we want to:</p>
<ol style="list-style-type: decimal">
<li>Estimate the branch length (shared genetic drift) between a pair of
populations <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span> with respect to a common outgroup <span class="math inline">\(C\)</span>. In this case, the higher the <span class="math inline">\(f_3\)</span> value, the longer the shared
evolutionary time between <span class="math inline">\(A\)</span> and
<span class="math inline">\(B\)</span>.</li>
<li>Test whether population <span class="math inline">\(C\)</span> is a
mixture of two populations <span class="math inline">\(A\)</span> and
<span class="math inline">\(B\)</span>. Significantly negative values of
the <span class="math inline">\(f_3\)</span> statistic are then a
statistical evidence of this admixture.</li>
</ol>
<p>As an example, imagine we are interested in relative divergence times
between pairs of present-day human populations, and want to know in
which approximate order they split of from each other. To address this
problem, we could use <span class="math inline">\(f_3\)</span> statistic
by fixing the <span class="math inline">\(C\)</span> outgroup as San,
and calculating pairwise <span class="math inline">\(f_3\)</span>
statistics between all present-day modern humans.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pops</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"French"</span>, <span class="st">"Sardinian"</span>, <span class="st">"Han"</span>, <span class="st">"Papuan"</span>, <span class="st">"Mbuti"</span>, <span class="st">"Dinka"</span>, <span class="st">"Yoruba"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/f4ratio.html">f3</a></span><span class="op">(</span>A <span class="op">=</span> <span class="va">pops</span>, B <span class="op">=</span> <span class="va">pops</span>, C <span class="op">=</span> <span class="st">"Khomani_San"</span>, data <span class="op">=</span> <span class="va">snps</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">A</th>
<th align="left">B</th>
<th align="left">C</th>
<th align="right">f3</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
<th align="right">nsnps</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">French</td>
<td align="left">French</td>
<td align="left">Khomani_San</td>
<td align="right">0.000000</td>
<td align="right">-1.000000</td>
<td align="right">0.000</td>
<td align="right">-1</td>
</tr>
<tr class="even">
<td align="left">French</td>
<td align="left">Sardinian</td>
<td align="left">Khomani_San</td>
<td align="right">0.353447</td>
<td align="right">0.012527</td>
<td align="right">28.215</td>
<td align="right">249760</td>
</tr>
<tr class="odd">
<td align="left">French</td>
<td align="left">Han</td>
<td align="left">Khomani_San</td>
<td align="right">0.316964</td>
<td align="right">0.011914</td>
<td align="right">26.604</td>
<td align="right">253158</td>
</tr>
<tr class="even">
<td align="left">French</td>
<td align="left">Papuan</td>
<td align="left">Khomani_San</td>
<td align="right">0.306962</td>
<td align="right">0.011708</td>
<td align="right">26.218</td>
<td align="right">251648</td>
</tr>
<tr class="odd">
<td align="left">French</td>
<td align="left">Mbuti</td>
<td align="left">Khomani_San</td>
<td align="right">0.119283</td>
<td align="right">0.008448</td>
<td align="right">14.119</td>
<td align="right">271501</td>
</tr>
<tr class="even">
<td align="left">French</td>
<td align="left">Dinka</td>
<td align="left">Khomani_San</td>
<td align="right">0.190141</td>
<td align="right">0.010049</td>
<td align="right">18.922</td>
<td align="right">276964</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sort the population labels according to an increasing f3 value relative to French</span></span>
<span><span class="va">ordered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">result</span>, <span class="va">A</span> <span class="op">==</span> <span class="st">"Mbuti"</span>, <span class="va">B</span> <span class="op">!=</span> <span class="st">"Mbuti"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">f3</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">.</span><span class="op">[[</span><span class="st">"B"</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Mbuti"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot heatmap of pairwise f3 values</span></span>
<span><span class="va">result</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">A</span> <span class="op">!=</span> <span class="va">B</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>A <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">A</span>, levels <span class="op">=</span> <span class="va">ordered</span><span class="op">)</span>,</span>
<span>         B <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">B</span>, levels <span class="op">=</span> <span class="va">ordered</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">A</span>, <span class="va">B</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">f3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="01-tutorial_files/figure-html/f3_plot-1.png" width="768"></p>
<p>We can see that when we order the heatmap labels based on values of
pairwise <span class="math inline">\(f_3\)</span> statistics, the
(already known) order of population splits pops up nicely (i.e. San
separated first, followed by Mbuti, etc.).</p>
</div>
<div class="section level2">
<h2 id="qpwave-and-qpadm">qpWave and qpAdm<a class="anchor" aria-label="anchor" href="#qpwave-and-qpadm"></a>
</h2>
<p>Both <em>qpWave</em> and <em>qpAdm</em> can be though of as more
complex and powerful extensions of the basic ideas behind a simple <span class="math inline">\(f_4\)</span> statistic. Building upon the <span class="math inline">\(f_4\)</span> theory and generalizing it,
<em>qpWave</em> makes it possible to find the lowest number of “streams
of ancestry” between two groups of populations that is consistent with
the data. Extending the concept of <span class="math inline">\(f_4\)</span> statistics even further,
<em>qpAdm</em> allows to find the proportions of ancestry from a set of
ancestral populations that contributed ancestry to our population of
interest.</p>
<p>Unfortunately, both methods represent a rather advanced topic that
still lacks proper documentation and beginner-friendly tutorials, and
explaining them in detail is beyond the scope of this vignette. If you
want to use them, it’s crucial that you read the official documentation
decribing the basic ideas of both methods (<a href="https://github.com/DReichLab/AdmixTools/blob/master/pdoc.pdf" class="external-link">distributed
with ADMIXTOOLS</a>), <em>and</em> that you read the relevant
supplementary sections of papers published by David Reich’s group. At
the very least, I recommend reading:</p>
<ul>
<li><p>Note S6 of <em>“<a href="https://www.nature.com/articles/nature11258" class="external-link">Reconstructing Native
American population history</a>”</em> by Reich et al. This paper first
introduced the theoretical background of what later became
<em>qpWave</em>.</p></li>
<li><p>Supplementary Information 10 of <em>“<a href="https://www.nature.com/articles/nature14317" class="external-link">Massive migration
from the steppe was a source for Indo-European languages in
Europe</a>”</em> by Haak et al., which gives a more consise overview of
the <em>qpWave</em> method than S6 of Reich et al. 2012, and also
introduces the <em>qpAdm</em> methodology for estimating admixture
proportions.</p></li>
<li><p>A phenomenal description of <em>qpAdm</em> methodology and best
practices by <a href="https://www.biorxiv.org/content/10.1101/2020.04.09.032664v1" class="external-link">Harney
<em>et al.</em> 2020</a> (see also PDF with <a href="https://www.biorxiv.org/content/biorxiv/early/2020/04/10/2020.04.09.032664/DC1/embed/media-1.pdf?download=true" class="external-link">practical
guidelines</a>. Both documents should really be a mandatory reading
before doing any <em>qpAdm</em> analysis.</p></li>
</ul>
<p>In the remainder of this section, I will assume that you are familiar
with both methods, and will only explain how to use <em>admixr</em> for
running them from R.</p>
<div class="section level3">
<h3 id="qpwave">
<em>qpWave</em><a class="anchor" aria-label="anchor" href="#qpwave"></a>
</h3>
<p>To run <em>qpWave</em>, you must provide a list of <em>left</em> and
<em>right</em> populations (using the terminology of Haak et al. 2015
above). The aim of the method is to get an idea about the number of
migration waves from <em>right</em> to <em>left</em> (with no
back-migration from <em>left</em> to <em>right</em>!). This is done by
estimating the rank of a matrix of all possible <span class="math inline">\(f_4\)</span> statistics</p>
<p><span class="math display">\[f_4(\textrm{left}_1, \textrm{left}_i;
\textrm{right}_1, \textrm{right}_i),\]</span></p>
<p>where <span class="math inline">\(\textrm{left}_1\)</span> and <span class="math inline">\(\textrm{right}_1\)</span> are some fixed
populations and the <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> indices run over all other possible
choices of populations.</p>
<p>As an example, let’s try to find the number of admixture waves from
<em>right</em> = {Yoruba, Mbuti, Alta} into <em>left</em> = {French,
Sardinian, Han} populations. We can do this using the function
<code><a href="../reference/qpWave.html">qpWave()</a></code>, setting its arguments appropriately:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qpWave.html">qpWave</a></span><span class="op">(</span></span>
<span> left <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"French"</span>, <span class="st">"Sardinian"</span>, <span class="st">"Han"</span><span class="op">)</span>,</span>
<span> right <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Altai"</span>, <span class="st">"Yoruba"</span>, <span class="st">"Mbuti"</span><span class="op">)</span>,</span>
<span> data <span class="op">=</span> <span class="va">snps</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/qpWave.html">qpWave()</a></code> function returns a data frame which shows
the results of a series of matrix rank tests. The <code>rank</code>
column is the matrix rank tested, <code>df</code>, <code>chisq</code>
and <code>tail</code> give the degrees of freedom, <span class="math inline">\(\chi^2\)</span> value and <span class="math inline">\(p\)</span>-value for the comparison with the
saturated model (the <span class="math inline">\(p\)</span>-value then
indicates which matrix rank is consistent with the data - see example
below), and <code>dfdiff</code>, <code>chisqdiff</code> and
<code>taildiff</code> give the same, but always comparing a model to the
model with one rank less.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="right">rank</th>
<th align="right">df</th>
<th align="right">chisq</th>
<th align="right">tail</th>
<th align="right">dfdiff</th>
<th align="right">chisqdiff</th>
<th align="right">taildiff</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">0</td>
<td align="right">4</td>
<td align="right">1.756</td>
<td align="right">0.7805106</td>
<td align="right">0</td>
<td align="right">0.000</td>
<td align="right">1.0000000</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0.192</td>
<td align="right">0.6614968</td>
<td align="right">3</td>
<td align="right">1.564</td>
<td align="right">0.6674998</td>
</tr>
<tr class="odd">
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0.000</td>
<td align="right">1.0000000</td>
<td align="right">1</td>
<td align="right">0.192</td>
<td align="right">0.6614968</td>
</tr>
</tbody>
</table>
<p>In this example, we see that matrix <span class="math inline">\(r =
0\)</span> cannot be rejected (<code>tail</code> <span class="math inline">\(p\)</span>-value = 0.78). Because Reich et
al. 2012 showed that <span class="math inline">\(r + 1 \le n\)</span>,
where <span class="math inline">\(n\)</span> is the number of admixture
waves, we can interpret this as <em>left</em> populations having at
least <span class="math inline">\(n = 1\)</span> streams of ancestry
from the set of <em>right</em> populations. In this case, the most
likely explanation is Neandertal admixture into non-Africans today.</p>
<p>Now, what happens if we add Papuans to the <em>left</em> group?</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qpWave.html">qpWave</a></span><span class="op">(</span></span>
<span> left <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Papuan"</span>, <span class="st">"French"</span>, <span class="st">"Sardinian"</span>, <span class="st">"Han"</span><span class="op">)</span>,</span>
<span> right <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Altai"</span>, <span class="st">"Yoruba"</span>, <span class="st">"Mbuti"</span><span class="op">)</span>,</span>
<span> data <span class="op">=</span> <span class="va">snps</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="right">rank</th>
<th align="right">df</th>
<th align="right">chisq</th>
<th align="right">tail</th>
<th align="right">dfdiff</th>
<th align="right">chisqdiff</th>
<th align="right">taildiff</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">0</td>
<td align="right">6</td>
<td align="right">29.143</td>
<td align="right">0.0000572</td>
<td align="right">0</td>
<td align="right">0.000</td>
<td align="right">1.0000000</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="right">2</td>
<td align="right">0.601</td>
<td align="right">0.7403643</td>
<td align="right">4</td>
<td align="right">28.542</td>
<td align="right">0.0000097</td>
</tr>
<tr class="odd">
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0.000</td>
<td align="right">1.0000000</td>
<td align="right">2</td>
<td align="right">0.601</td>
<td align="right">0.7403643</td>
</tr>
</tbody>
</table>
<p>We can now clearly reject rank <span class="math inline">\(r =
0\)</span>, but we see that the data is consistent with rank <span class="math inline">\(r = 1\)</span>, meaning that there must have been
at least <span class="math inline">\(n = 2\)</span> streams of ancestry
from <em>right</em> to <em>left</em> populations (<span class="math inline">\(r + 1 \le n\)</span>). Because this happened after
we introduced Papuans to the <em>left</em> set, this could indicate a
separate pulse of archaic introgression into Papuans, which is not
surprising given what we know about significantly more archaic ancestry
in Papuans than in any other present-day population.</p>
</div>
<div class="section level3">
<h3 id="qpadm">
<em>qpAdm</em><a class="anchor" aria-label="anchor" href="#qpadm"></a>
</h3>
<p>The <em>qpAdm</em> method can be used to find, for a given target
population, the proportions of ancestry coming from a set of
<em>source</em> populations. Importantly, since we often lack accurate
representatives of the true ancestral populations, we can use a set of
reference populations instead, under a crucial assumption that the
references set is phylogenetically closer to true <em>source</em>
populations than to a set of specified <em>outgroups</em>. For example,
coming back to our example of estimating the proportions of Neandertal
ancestry in people today, we could define:</p>
<ul>
<li>a set of European individuals as the <em>target</em>;</li>
<li>Vindija Neanderthal and an African as two <em>source</em>
populations;</li>
<li>
<em>outgroup</em> populations as Chimp, Altai Neanderthal and
Denisovan (which are all further from the true ancestral populations
than the specified <em>sources</em>).</li>
</ul>
<p>Having defined all three population sets, we can run qpAdm with:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qpAdm.html">qpAdm</a></span><span class="op">(</span></span>
<span>  target <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sardinian"</span>, <span class="st">"Han"</span>, <span class="st">"French"</span><span class="op">)</span>,</span>
<span>  sources <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Vindija"</span>, <span class="st">"Yoruba"</span><span class="op">)</span>,</span>
<span>  outgroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Chimp"</span>, <span class="st">"Denisova"</span>, <span class="st">"Altai"</span><span class="op">)</span>,</span>
<span>  data <span class="op">=</span> <span class="va">snps</span>,</span>
<span>  params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>inbreed <span class="op">=</span> <span class="st">"YES"</span><span class="op">)</span> <span class="co"># forced by new ADMIXTOOLS qpfstats</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/qpAdm.html">qpAdm()</a></code> function has an argument
<code>details</code> (default TRUE) which makes the function return a
list of three elements:</p>
<ul>
<li>
<code>proportions</code> - data frame with admixture proportions -
this is what we mostly care about;</li>
<li>
<code>ranks</code> - results of rank tests performed by
<em>qpWave</em> - these evaluate how well does the assumed
traget-sources-outgroups population model match the data;</li>
<li>
<code>subsets</code> - results of the “all subsets” analysis (see
the <a href="https://github.com/DReichLab/AdmixTools/blob/master/pdoc.pdf" class="external-link">documentation</a>
for more details.</li>
</ul>
<p>If <code>details</code> is set to <code>FALSE</code>, only the
<code>proportions</code> components is returned by the
<code><a href="../reference/qpAdm.html">qpAdm()</a></code> function.</p>
<p>Let’s start with the <code>ranks</code> element:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span><span class="op">$</span><span class="va">ranks</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">target</th>
<th align="right">rank</th>
<th align="right">df</th>
<th align="right">chisq</th>
<th align="right">tail</th>
<th align="right">dfdiff</th>
<th align="right">chisqdiff</th>
<th align="right">taildiff</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Sardinian</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0.006</td>
<td align="right">0.9362605</td>
<td align="right">3</td>
<td align="right">-0.006</td>
<td align="right">1.0000000</td>
</tr>
<tr class="even">
<td align="left">Sardinian</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0.000</td>
<td align="right">1.0000000</td>
<td align="right">1</td>
<td align="right">0.006</td>
<td align="right">0.9362605</td>
</tr>
<tr class="odd">
<td align="left">Han</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">2.144</td>
<td align="right">0.1431157</td>
<td align="right">3</td>
<td align="right">-2.144</td>
<td align="right">1.0000000</td>
</tr>
<tr class="even">
<td align="left">Han</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0.000</td>
<td align="right">1.0000000</td>
<td align="right">1</td>
<td align="right">2.144</td>
<td align="right">0.1431157</td>
</tr>
<tr class="odd">
<td align="left">French</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">3.814</td>
<td align="right">0.0508171</td>
<td align="right">3</td>
<td align="right">-3.814</td>
<td align="right">1.0000000</td>
</tr>
<tr class="even">
<td align="left">French</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0.000</td>
<td align="right">1.0000000</td>
<td align="right">1</td>
<td align="right">3.814</td>
<td align="right">0.0508171</td>
</tr>
</tbody>
</table>
<p>The row with rank = 1 represents a <em>qpWave</em> test with all
<span class="math inline">\(n\)</span> <em>source</em> populations set
as the <em>left</em> set and all <em>outgroups</em> as the
<em>right</em> set. This test evaluates whether the ancestral
populations are descended from <span class="math inline">\(n\)</span>
independent streams of ancestry. In our case, <span class="math inline">\(n = 2\)</span> (Mbuti and Vindija), which means
that the data would have to be consistent with rank <span class="math inline">\(r = 1\)</span> to satisfy the inequality <span class="math inline">\(r + 1 \le n\)</span> proved by Reich et al., 2012.
We see that this is true for all three target populations (<span class="math inline">\(p\)</span>-value &gt; 0.05 for all targets), and
the simple model of Neandertal admixture thus seems to be reasonably
consistent with the data.</p>
<p>The rank = 2 row represents a <em>qpWave</em> test after adding a
target population to the <em>left</em> group together with the
<em>sources</em>. This test makes sure that including the target
population does not increase the rank of the <span class="math inline">\(f_4\)</span> matrix, meaning that the target can
be really modelled as a mixture of ancestries from the <em>sources</em>.
If the <span class="math inline">\(p\)</span>-values turn out to be very
low, this indicates that the assumed model does not fit the data and
that a part of the ancestry in a <em>target</em> possibly cannot be
traced to any of the <em>sources</em>. In our case, however, all rank =
2 test <span class="math inline">\(p\)</span>-values are not
significant, and we can be reasonably sure that the <em>target</em>
samples can be fully modelled as a mixtures of all specified
<em>references</em>.</p>
<p>The most important element of a <em>qpAdm</em> output is in the
<code>$proportions</code> component. This contains admixture proportion
estimates from all specified sources, the p-values of each model
(remember, low p-values/significance means the model is rejected!) as
well as standard errors for those proportions using a block
jackknife:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span><span class="op">$</span><span class="va">proportions</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="11%">
<col width="9%">
<col width="7%">
<col width="17%">
<col width="15%">
<col width="12%">
<col width="14%">
<col width="11%">
</colgroup>
<thead><tr class="header">
<th align="left">target</th>
<th align="right">Vindija</th>
<th align="right">Yoruba</th>
<th align="right">stderr_Vindija</th>
<th align="right">stderr_Yoruba</th>
<th align="right">nsnps_used</th>
<th align="right">nsnps_target</th>
<th align="right">pvalue</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Sardinian</td>
<td align="right">0.025</td>
<td align="right">0.975</td>
<td align="right">0.006</td>
<td align="right">0.006</td>
<td align="right">492420</td>
<td align="right">499314</td>
<td align="right">0.9362610</td>
</tr>
<tr class="even">
<td align="left">Han</td>
<td align="right">0.021</td>
<td align="right">0.979</td>
<td align="right">0.006</td>
<td align="right">0.006</td>
<td align="right">492531</td>
<td align="right">499654</td>
<td align="right">0.1431160</td>
</tr>
<tr class="odd">
<td align="left">French</td>
<td align="right">0.022</td>
<td align="right">0.978</td>
<td align="right">0.006</td>
<td align="right">0.006</td>
<td align="right">492490</td>
<td align="right">499434</td>
<td align="right">0.0508171</td>
</tr>
</tbody>
</table>
<p>If we compare this result to the <span class="math inline">\(f_4\)</span>-ratio values calculated above, we see
that the <em>qpAdm</em> estimates are very close to what we got
earlier.</p>
<p>The third element in the list of results shows the outcome of an “all
subsets” analysis, which involves testing all subsets of potential
source populations. Each 1 in the “pattern” column means that the
proportion of ancestry from that particular source population (in the
order as specified by the user) was forced to 0.0.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span><span class="op">$</span><span class="va">subsets</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">target</th>
<th align="left">pattern</th>
<th align="right">wt</th>
<th align="right">dof</th>
<th align="right">chisq</th>
<th align="right">tail</th>
<th align="right">Vindija</th>
<th align="right">Yoruba</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Sardinian</td>
<td align="left">00</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0.006</td>
<td align="right">0.9362610</td>
<td align="right">0.025</td>
<td align="right">0.975</td>
</tr>
<tr class="even">
<td align="left">Sardinian</td>
<td align="left">01</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">15953.171</td>
<td align="right">0.0000000</td>
<td align="right">1.000</td>
<td align="right">0.000</td>
</tr>
<tr class="odd">
<td align="left">Sardinian</td>
<td align="left">10</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">16.564</td>
<td align="right">0.0002530</td>
<td align="right">0.000</td>
<td align="right">1.000</td>
</tr>
<tr class="even">
<td align="left">Han</td>
<td align="left">00</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">2.144</td>
<td align="right">0.1431160</td>
<td align="right">0.021</td>
<td align="right">0.979</td>
</tr>
<tr class="odd">
<td align="left">Han</td>
<td align="left">01</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">14965.791</td>
<td align="right">0.0000000</td>
<td align="right">1.000</td>
<td align="right">0.000</td>
</tr>
<tr class="even">
<td align="left">Han</td>
<td align="left">10</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">14.454</td>
<td align="right">0.0007269</td>
<td align="right">0.000</td>
<td align="right">1.000</td>
</tr>
<tr class="odd">
<td align="left">French</td>
<td align="left">00</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">3.814</td>
<td align="right">0.0508171</td>
<td align="right">0.022</td>
<td align="right">0.978</td>
</tr>
<tr class="even">
<td align="left">French</td>
<td align="left">01</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">15441.258</td>
<td align="right">0.0000000</td>
<td align="right">1.000</td>
<td align="right">0.000</td>
</tr>
<tr class="odd">
<td align="left">French</td>
<td align="left">10</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">16.028</td>
<td align="right">0.0003308</td>
<td align="right">0.000</td>
<td align="right">1.000</td>
</tr>
</tbody>
</table>
<p><em>New feature</em>: There is a new function called
<code><a href="../reference/qpAdm_rotation.html">qpAdm_rotation()</a></code> which allows exhaustive exploration of
many <em>qpAdm</em> models in paralle. For more information please see
the vignette _“Fitting qpAdm models with a ‘rotation’ strategy_”.</p>
</div>
</div>
<div class="section level2">
<h2 id="grouping-samples">Grouping samples<a class="anchor" aria-label="anchor" href="#grouping-samples"></a>
</h2>
<p>What we’ve been doing so far was calculating statistics for
individual samples. However, it is often useful to treat multiple
samples as a single group or population. <em>admixr</em> provides a
function called <code><a href="../reference/relabel.html">relabel()</a></code> that does just that.</p>
<p>Here is an example: let’s say we want to run a similar analysis to
the one described in the <span class="math inline">\(D\)</span>
statistic section, but we want to treat Europeans, Africans and archaics
as combined populations, and not as separate individuals. But the
<code>ind</code> file that we have does not contain grouped labels -
each sample stands on its own:</p>
<pre><code>Chimp        U  Chimp
Mbuti        U  Mbuti
Yoruba       U  Yoruba
Khomani_San  U  Khomani_San
Han          U  Han
Dinka        U  Dinka
Sardinian    U  Sardinian
Papuan       U  Papuan
French       U  French
Vindija      U  Vindija
Altai        U  Altai
Denisova     U  Denisova</code></pre>
<p>To merge several individual samples under a combined label we can
call <code><a href="../reference/relabel.html">relabel()</a></code> like this:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># paths to the original ind file and a new modified ind file, which will</span></span>
<span><span class="co"># contain merged population labels</span></span>
<span><span class="va">modif_snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/relabel.html">relabel</a></span><span class="op">(</span></span>
<span>  <span class="va">snps</span>,</span>
<span>  European <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"French"</span>, <span class="st">"Sardinian"</span><span class="op">)</span>,</span>
<span>  African <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Dinka"</span>, <span class="st">"Yoruba"</span>, <span class="st">"Mbuti"</span>, <span class="st">"Khomani_San"</span><span class="op">)</span>,</span>
<span>  Archaic <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Vindija"</span>, <span class="st">"Altai"</span>, <span class="st">"Denisova"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">modif_snps</span></span>
<span><span class="co">#&gt; EIGENSTRAT object</span></span>
<span><span class="co">#&gt; =================</span></span>
<span><span class="co">#&gt; components:</span></span>
<span><span class="co">#&gt;   ind file: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.ind</span></span>
<span><span class="co">#&gt;   snp file: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.snp</span></span>
<span><span class="co">#&gt;   geno file: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.geno</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; modifiers:</span></span>
<span><span class="co">#&gt;   groups:  /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/file5cb6473bb754.ind</span></span></code></pre></div>
<p>We can see that the function <code>relabel</code> returned a modified
<code>EIGENSTRAT</code> object, which contains a new item in the
“modifiers” section - the path to a new ind file. Let’s look at its
contents:</p>
<pre><code>Chimp        U  Chimp
Mbuti        U  African
Yoruba       U  African
Khomani_San  U  African
Han          U  Han
Dinka        U  African
Sardinian    U  European
Papuan       U  Papuan
French       U  European
Vindija      U  Archaic
Altai        U  Archaic
Denisova     U  Archaic</code></pre>
<p>Having the modified <code>EIGENSTRAT</code> object ready, we can then
use “European”, “African” and “Archaic” names in any of the
<em>admixr</em> wrapper functions described above. For example:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/f4ratio.html">d</a></span><span class="op">(</span>W <span class="op">=</span> <span class="st">"European"</span>, X <span class="op">=</span> <span class="st">"African"</span>, Y <span class="op">=</span> <span class="st">"Archaic"</span>, Z <span class="op">=</span> <span class="st">"Chimp"</span>, data <span class="op">=</span> <span class="va">modif_snps</span><span class="op">)</span></span></code></pre></div>
<p>Here is the result, showing again Europeans show genetic affinity to
archaic humans compared to Africans today:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="12%">
<col width="10%">
<col width="10%">
<col width="8%">
<col width="9%">
<col width="12%">
<col width="9%">
<col width="8%">
<col width="8%">
<col width="9%">
</colgroup>
<thead><tr class="header">
<th align="left">W</th>
<th align="left">X</th>
<th align="left">Y</th>
<th align="left">Z</th>
<th align="right">D</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
<th align="right">BABA</th>
<th align="right">ABBA</th>
<th align="right">nsnps</th>
</tr></thead>
<tbody><tr class="odd">
<td align="left">European</td>
<td align="left">African</td>
<td align="left">Archaic</td>
<td align="left">Chimp</td>
<td align="right">0.0225</td>
<td align="right">0.004404</td>
<td align="right">5.117</td>
<td align="right">15487</td>
<td align="right">14805</td>
<td align="right">489003</td>
</tr></tbody>
</table>
<p>Note that the <code><a href="../reference/f4ratio.html">d()</a></code> function correctly picks up the “group
modifier” <code>ind</code> file from the provided
<code>EIGENSTRAT</code> object and uses it in place of the original
<code>ind</code> file.</p>
</div>
<div class="section level2">
<h2 id="counting-presentmissing-snps">Counting present/missing SNPs<a class="anchor" aria-label="anchor" href="#counting-presentmissing-snps"></a>
</h2>
<p>The <code>count_snps</code> function can be useful for quality
control, weighting of admixture statistics (<span class="math inline">\(D\)</span>, <span class="math inline">\(f_4\)</span>, etc.) in regression analyses etc.
There are two optional arguments:</p>
<ul>
<li>
<code>prop</code> - changes whether to report SNP counts or
proportions (set to <code>FALSE</code> by default),</li>
<li>
<code>missing</code> - controls whether to count missing SNPs
instead of present SNPs (set to <code>FALSE</code> by default).</li>
</ul>
<p>For each sample, count the SNPs present in that sample:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/count_snps.html">count_snps</a></span><span class="op">(</span><span class="va">snps</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">id</th>
<th align="left">sex</th>
<th align="left">label</th>
<th align="right">present</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Chimp</td>
<td align="left">U</td>
<td align="left">Chimp</td>
<td align="right">491273</td>
</tr>
<tr class="even">
<td align="left">Mbuti</td>
<td align="left">U</td>
<td align="left">Mbuti</td>
<td align="right">499334</td>
</tr>
<tr class="odd">
<td align="left">Yoruba</td>
<td align="left">U</td>
<td align="left">Yoruba</td>
<td align="right">499246</td>
</tr>
<tr class="even">
<td align="left">Khomani_San</td>
<td align="left">U</td>
<td align="left">Khomani_San</td>
<td align="right">499250</td>
</tr>
<tr class="odd">
<td align="left">Han</td>
<td align="left">U</td>
<td align="left">Han</td>
<td align="right">499654</td>
</tr>
<tr class="even">
<td align="left">Dinka</td>
<td align="left">U</td>
<td align="left">Dinka</td>
<td align="right">499362</td>
</tr>
<tr class="odd">
<td align="left">Sardinian</td>
<td align="left">U</td>
<td align="left">Sardinian</td>
<td align="right">499314</td>
</tr>
<tr class="even">
<td align="left">Papuan</td>
<td align="left">U</td>
<td align="left">Papuan</td>
<td align="right">499377</td>
</tr>
<tr class="odd">
<td align="left">French</td>
<td align="left">U</td>
<td align="left">French</td>
<td align="right">499434</td>
</tr>
<tr class="even">
<td align="left">Vindija</td>
<td align="left">U</td>
<td align="left">Vindija</td>
<td align="right">497544</td>
</tr>
<tr class="odd">
<td align="left">Altai</td>
<td align="left">U</td>
<td align="left">Altai</td>
<td align="right">497729</td>
</tr>
<tr class="even">
<td align="left">Denisova</td>
<td align="left">U</td>
<td align="left">Denisova</td>
<td align="right">497398</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="data-filtering">Data filtering<a class="anchor" aria-label="anchor" href="#data-filtering"></a>
</h2>
<div class="section level3">
<h3 id="filtering-based-on-a-bed-file">Filtering based on a BED file<a class="anchor" aria-label="anchor" href="#filtering-based-on-a-bed-file"></a>
</h3>
<p>It is quite common to repeat a particular analysis only on a subset
of the genome (such as intergenic sites, etc). However, EIGENSTRAT is a
rather obscure file format which is generally not supported by standard
bioinformatics tools. Luckily, <em>admixr</em> includes a function
<code><a href="../reference/filter_bed.html">filter_bed()</a></code> that takes an <code>EIGENSTRAT</code> object
and a BED file as its inputs and produces a new object that contains a
modifier called “excluded”, linking to a snp file with coordinates of
sites that did not pass the filtering and will be excluded from later
analyses.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">dirname</a></span><span class="op">(</span><span class="va">prefix</span><span class="op">)</span>, <span class="st">"regions.bed"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># BED file contains regions to keep in an analysis</span></span>
<span><span class="va">new_snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_bed.html">filter_bed</a></span><span class="op">(</span><span class="va">snps</span>, <span class="va">bed</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># BED file contains regions to remove from an analysis</span></span>
<span><span class="va">new_snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_bed.html">filter_bed</a></span><span class="op">(</span><span class="va">snps</span>, <span class="va">bed</span>, remove <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new_snps</span></span>
<span><span class="co">#&gt; EIGENSTRAT object</span></span>
<span><span class="co">#&gt; =================</span></span>
<span><span class="co">#&gt; components:</span></span>
<span><span class="co">#&gt;   ind file: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.ind</span></span>
<span><span class="co">#&gt;   snp file: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.snp</span></span>
<span><span class="co">#&gt;   geno file: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.geno</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; modifiers:</span></span>
<span><span class="co">#&gt;   excluded sites:  /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/file5cb6227326a3.snp </span></span>
<span><span class="co">#&gt;       (SNPs excluded: 100000, SNPs remaining: 400000)</span></span></code></pre></div>
<p>If we want to run the whole analysis in a single pipeline, we can use
the <code>%&gt;%</code> pipe operator and do the following:</p>
<p>(The <code>%&gt;%</code> operator takes what is on its left side and
puts it as a first argument of a function on the right side. While it
takes some time to get used to, it is very useful in longer multi-step
“pipelines” because it makes more pipelines much more readable. In fact,
the resulting code often reads <em>almost</em> like English! The
<code>%&gt;%</code> pipe is automatically imported when you load the
<code>tidyverse</code> library, and you can read about it more <a href="https://magrittr.tidyverse.org" class="external-link">here</a>.)</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/filter_bed.html">filter_bed</a></span><span class="op">(</span><span class="st">"regions.bed"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/f4ratio.html">d</a></span><span class="op">(</span>W <span class="op">=</span> <span class="st">"French"</span>, X <span class="op">=</span> <span class="st">"Mbuti"</span>, Y <span class="op">=</span> <span class="st">"Vindija"</span>, Z <span class="op">=</span> <span class="st">"Chimp"</span><span class="op">)</span></span></code></pre></div>
<p>This is because in the formal definitions of <em>admixr</em>
function, <code>data =</code> is always the argument, so we don’t have
to specify it manually.</p>
<p><strong>Important:</strong> The <code><a href="../reference/filter_bed.html">filter_bed()</a></code> function
makes it very easy to do filtering without worrying about locations of
intermediate files, but it is important to keep in mind that the
function still creates temporary files under the hood. If you plan to
run many independent calculations on a filtered subset of the data, it’s
better to save the new <code>EIGENSTRAT</code> object to a variable
first and re-use the same object multiple times, rather than running the
whole pipeline for each analysis separately (which would create new
copies of intermediate files for each iteration).</p>
</div>
<div class="section level3">
<h3 id="filtering-out-potential-ancient-dna-damage-snps">Filtering out potential ancient DNA damage SNPs<a class="anchor" aria-label="anchor" href="#filtering-out-potential-ancient-dna-damage-snps"></a>
</h3>
<p>In the field of ancient DNA, we often need to repeat an analysis on a
subset of data that is less likely to be influenced by ancient DNA
damage, to verify that our results are not caused by artifacts in the
data (due to biochemical properties of DNA degradation, ancient DNA
damage will lead to an increase in C→T and G→A substitutions). Using a
similar method described in the BED filtering section above, we can use
the <code><a href="../reference/transversions_only.html">transversions_only()</a></code> function to generate a snp file
with positions that carry transitions (C→T and G→A sites):</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new_snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transversions_only.html">transversions_only</a></span><span class="op">(</span><span class="va">snps</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># perform the calculation only on transversions</span></span>
<span><span class="fu"><a href="../reference/f4ratio.html">d</a></span><span class="op">(</span>W <span class="op">=</span> <span class="st">"French"</span>, X <span class="op">=</span> <span class="st">"Dinka"</span>, Y <span class="op">=</span> <span class="st">"Altai"</span>, Z <span class="op">=</span> <span class="st">"Chimp"</span>, data <span class="op">=</span> <span class="va">new_snps</span><span class="op">)</span></span></code></pre></div>
<p>Again, we could combine several filtering steps into one
pipeline:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>                                    <span class="co"># take the original data</span></span>
<span>  <span class="fu"><a href="../reference/filter_bed.html">filter_bed</a></span><span class="op">(</span><span class="st">"regions.bed"</span>, remove <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>  <span class="co"># remove sites not in specified regions</span></span>
<span>  <span class="fu"><a href="../reference/transversions_only.html">transversions_only</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>                      <span class="co"># remove potential false SNPs due to aDNA damage</span></span>
<span>  <span class="fu"><a href="../reference/f4ratio.html">d</a></span><span class="op">(</span>W <span class="op">=</span> <span class="st">"French"</span>, X <span class="op">=</span> <span class="st">"Dinka"</span>, Y <span class="op">=</span> <span class="st">"Altai"</span>, Z <span class="op">=</span> <span class="st">"Chimp"</span><span class="op">)</span> <span class="co"># calculate D on the filtered dataset</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="merging-eigenstrat-datasets">Merging EIGENSTRAT datasets<a class="anchor" aria-label="anchor" href="#merging-eigenstrat-datasets"></a>
</h2>
<p>Another useful data processing function is
<code><a href="../reference/merge_eigenstrat.html">merge_eigenstrat()</a></code>. This function takes two EIGENSTRAT
datasets and merges them, producing a union of samples and intersection
of SNPs from both of them and returning a new <code>EIGENSTRAT</code>
object.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="co"># this is just an example code - it will not run unless you specify the paths</span></span>
<span id="cb44-2"><a href="#cb44-2" tabindex="-1"></a>merged <span class="ot">&lt;-</span> <span class="fu">merge_eigenstrat</span>(</span>
<span id="cb44-3"><a href="#cb44-3" tabindex="-1"></a>    <span class="at">merged =</span> <span class="sc">&lt;</span><span class="st">"prefix of the merged dataset"</span><span class="sc">&gt;</span></span>
<span id="cb44-4"><a href="#cb44-4" tabindex="-1"></a>    <span class="at">a =</span> first_EIGENSTRAT_object,</span>
<span id="cb44-5"><a href="#cb44-5" tabindex="-1"></a>    <span class="at">b =</span> second_EIGENSTRAT_object</span>
<span id="cb44-6"><a href="#cb44-6" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="examining-log-information">Examining log information<a class="anchor" aria-label="anchor" href="#examining-log-information"></a>
</h2>
<p>The goal of <em>admixr</em> is to abstract away all the low-level
technical workings of ADMIXTOOLS. As we saw in the examples above, it
achieves this by doing all the dirty work of parsing the output files
generated by ADMIXTOOLS, presenting the user with convenient R data
structures.</p>
<p>Nevertheless, admixr cannot (yet) parse <em>all</em> the information
generated by ADMIXTOOLS utilities. For calculating D statistics, <span class="math inline">\(f_4\)</span> statistics and doing other simple
analyses, a single data frame is usually all the user needs. However,
sometimes things go wrong - it turns out data was in a wrong format, or
ADMIXTOOLS crashes with an error or the results simply look suspicious.
Furthermore, some complex commands (such as <em>qpAdm</em>) are not yet
completely implemented in admixr. In any of these cases, it is still
useful (and important!) to examine the log outputs.</p>
<p>Each admixr result, such as the data frame object below, contains a
(hidden!) attribute which carries the complete log file associate with
that particular analysis:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dres</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/f4ratio.html">d</a></span><span class="op">(</span>W <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"French"</span>, <span class="st">"Han"</span>, <span class="st">"Dinka"</span><span class="op">)</span>, X <span class="op">=</span> <span class="st">"Yoruba"</span>, Y <span class="op">=</span> <span class="st">"Vindija"</span>, Z <span class="op">=</span> <span class="st">"Chimp"</span>, data <span class="op">=</span> <span class="va">snps</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dres</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 10</span></span></span>
<span><span class="co">#&gt;   W      X      Y       Z           D  stderr Zscore  BABA  ABBA  nsnps</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> French Yoruba Vindija Chimp  0.031<span style="text-decoration: underline;">3</span> 0.006<span style="text-decoration: underline;">93</span>  4.51  <span style="text-decoration: underline;">15</span>802 <span style="text-decoration: underline;">14</span>844 <span style="text-decoration: underline;">487</span>753</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Han    Yoruba Vindija Chimp  0.027<span style="text-decoration: underline;">8</span> 0.006<span style="text-decoration: underline;">61</span>  4.20  <span style="text-decoration: underline;">15</span>780 <span style="text-decoration: underline;">14</span>928 <span style="text-decoration: underline;">487</span>925</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Dinka  Yoruba Vindija Chimp -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">9</span> 0.006<span style="text-decoration: underline;">12</span> -<span style="color: #BB0000;">0.151</span> <span style="text-decoration: underline;">15</span>131 <span style="text-decoration: underline;">15</span>159 <span style="text-decoration: underline;">487</span>667</span></span></code></pre></div>
<p>We can examine the full log of this D statistic run with the
<code>loginfo</code> function. Without any further arguments, this
function writes the complete log on the screen. Note that we run this
function <em>on the data frame object</em> although it might appear that
it does not contain any other information (it does but it’s normally
hidden as most of the time it’s not useful):</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/loginfo.html">loginfo</a></span><span class="op">(</span><span class="va">dres</span><span class="op">)</span></span>
<span><span class="co">#&gt; Full output log of qpDstat:</span></span>
<span><span class="co">#&gt; ===================================================</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; qpDstat: parameter file: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/qpDstat__335645284.par</span></span>
<span><span class="co">#&gt; ### THE INPUT PARAMETERS</span></span>
<span><span class="co">#&gt; ##PARAMETER NAME: VALUE</span></span>
<span><span class="co">#&gt; genotypename: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.geno</span></span>
<span><span class="co">#&gt; snpname: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.snp</span></span>
<span><span class="co">#&gt; indivname: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.ind</span></span>
<span><span class="co">#&gt; popfilename: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/qpDstat__335645284.pop</span></span>
<span><span class="co">#&gt; printsd: YES</span></span>
<span><span class="co">#&gt; ## qpDstat version: 980</span></span>
<span><span class="co">#&gt; inbreed set NO</span></span>
<span><span class="co">#&gt; number of quadruples 3</span></span>
<span><span class="co">#&gt;   0               French    1</span></span>
<span><span class="co">#&gt;   1                  Han    1</span></span>
<span><span class="co">#&gt;   2                Dinka    1</span></span>
<span><span class="co">#&gt;   3               Yoruba    1</span></span>
<span><span class="co">#&gt;   4              Vindija    1</span></span>
<span><span class="co">#&gt;   5                Chimp    1</span></span>
<span><span class="co">#&gt; jackknife block size:     0.050</span></span>
<span><span class="co">#&gt; snps: 500000  indivs: 6</span></span>
<span><span class="co">#&gt; number of blocks for block jackknife: 547</span></span>
<span><span class="co">#&gt; nrows, ncols: 6 500000</span></span>
<span><span class="co">#&gt; result:     French     Yoruba    Vindija      Chimp      0.0313      0.006933     4.510   15802  14844 487753 </span></span>
<span><span class="co">#&gt; result:        Han     Yoruba    Vindija      Chimp      0.0278      0.006609     4.199   15780  14928 487925 </span></span>
<span><span class="co">#&gt; result:      Dinka     Yoruba    Vindija      Chimp     -0.0009      0.006124    -0.151   15131  15159 487667 </span></span>
<span><span class="co">#&gt; ##end of qpDstat:        1.685 seconds cpu        0.000 Mbytes in use</span></span></code></pre></div>
<p>The situation is a bit more complex for <em>qpAdm</em> which can be
evaluated for multiple target populations/samples which are then
analyzed separately, with each analysis having its own log file:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qpadm_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qpAdm.html">qpAdm</a></span><span class="op">(</span></span>
<span>  target <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sardinian"</span>, <span class="st">"Han"</span><span class="op">)</span>,</span>
<span>  sources <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Vindija"</span>, <span class="st">"Yoruba"</span><span class="op">)</span>,</span>
<span>  outgroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Chimp"</span>, <span class="st">"Denisova"</span>, <span class="st">"Altai"</span><span class="op">)</span>,</span>
<span>  data <span class="op">=</span> <span class="va">snps</span>,</span>
<span>  params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>inbreed <span class="op">=</span> <span class="st">"YES"</span><span class="op">)</span> <span class="co"># forced by new ADMIXTOOLS qpfstats</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">qpadm_res</span></span>
<span><span class="co">#&gt; $proportions</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 8</span></span></span>
<span><span class="co">#&gt;   target    Vindija Yoruba stderr_Vindija stderr_Yoruba nsnps_used nsnps_target</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Sardinian   0.025  0.975          0.006         0.006     <span style="text-decoration: underline;">492</span>420       <span style="text-decoration: underline;">499</span>314</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Han         0.021  0.979          0.006         0.006     <span style="text-decoration: underline;">492</span>531       <span style="text-decoration: underline;">499</span>654</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: pvalue &lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $ranks</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 8</span></span></span>
<span><span class="co">#&gt;   target     rank    df chisq  tail dfdiff chisqdiff taildiff</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Sardinian     1     1 0.006 0.936      3    -<span style="color: #BB0000;">0.006</span>    1    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Sardinian     2     0 0     1          1     0.006    0.936</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Han           1     1 2.14  0.143      3    -<span style="color: #BB0000;">2.14</span>     1    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Han           2     0 0     1          1     2.14     0.143</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $subsets</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 8</span></span></span>
<span><span class="co">#&gt;   target    pattern    wt   dof     chisq     tail Vindija Yoruba</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Sardinian 00          0     1     0.006 0.936      0.025  0.975</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Sardinian 01          1     2 <span style="text-decoration: underline;">15</span>953.    0          1      0    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Sardinian 10          1     2    16.6   0.000<span style="text-decoration: underline;">253</span>   0      1    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Han       00          0     1     2.14  0.143      0.021  0.979</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Han       01          1     2 <span style="text-decoration: underline;">14</span>966.    0          1      0    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Han       10          1     2    14.5   0.000<span style="text-decoration: underline;">727</span>   0      1</span></span></code></pre></div>
<p>In this case, we could either print all log information (one log
output for each target we specified in the <em>qpAdm</em> call) by
running<code>loginfo(qpadm_res)</code> or, perhaps more useful, specify
which target’s log file we want to examine:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/loginfo.html">loginfo</a></span><span class="op">(</span><span class="va">qpadm_res</span>, target <span class="op">=</span> <span class="st">"Han"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Full output log of qpAdm for target 'Han':</span></span>
<span><span class="co">#&gt; ===================================================</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; qpAdm: parameter file: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/qpAdm__1789631171.par</span></span>
<span><span class="co">#&gt; ### THE INPUT PARAMETERS</span></span>
<span><span class="co">#&gt; ##PARAMETER NAME: VALUE</span></span>
<span><span class="co">#&gt; genotypename: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.geno</span></span>
<span><span class="co">#&gt; snpname: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.snp</span></span>
<span><span class="co">#&gt; indivname: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/snps/snps.ind</span></span>
<span><span class="co">#&gt; popleft: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/qpAdm__1789631171.popleft</span></span>
<span><span class="co">#&gt; popright: /var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmpzHxCwT/qpAdm__1789631171.popright</span></span>
<span><span class="co">#&gt; inbreed: YES</span></span>
<span><span class="co">#&gt; ## qpAdm version: 1520</span></span>
<span><span class="co">#&gt; seed: -1064939153</span></span>
<span><span class="co">#&gt; inbreed set YES</span></span>
<span><span class="co">#&gt; allsnps set NO.  It is recommended that allsnps be set explicitly</span></span>
<span><span class="co">#&gt; inbreed set YES</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; left pops:</span></span>
<span><span class="co">#&gt; Han</span></span>
<span><span class="co">#&gt; Vindija</span></span>
<span><span class="co">#&gt; Yoruba</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; right pops:</span></span>
<span><span class="co">#&gt; Chimp</span></span>
<span><span class="co">#&gt; Denisova</span></span>
<span><span class="co">#&gt; Altai</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   0                  Han    1</span></span>
<span><span class="co">#&gt;   1              Vindija    1</span></span>
<span><span class="co">#&gt;   2               Yoruba    1</span></span>
<span><span class="co">#&gt;   3                Chimp    1</span></span>
<span><span class="co">#&gt;   4             Denisova    1</span></span>
<span><span class="co">#&gt;   5                Altai    1</span></span>
<span><span class="co">#&gt; jackknife block size:     0.050</span></span>
<span><span class="co">#&gt; snps: 500000  indivs: 6</span></span>
<span><span class="co">#&gt; number of blocks for block jackknife: 547</span></span>
<span><span class="co">#&gt; ## ncols: 500000</span></span>
<span><span class="co">#&gt; coverage:                  Han 499654</span></span>
<span><span class="co">#&gt; coverage:              Vindija 497544</span></span>
<span><span class="co">#&gt; coverage:               Yoruba 499246</span></span>
<span><span class="co">#&gt; coverage:                Chimp 491273</span></span>
<span><span class="co">#&gt; coverage:             Denisova 497398</span></span>
<span><span class="co">#&gt; coverage:                Altai 497729</span></span>
<span><span class="co">#&gt; Effective number of blocks:   489.431</span></span>
<span><span class="co">#&gt; numsnps used: 492531</span></span>
<span><span class="co">#&gt; codimension 1</span></span>
<span><span class="co">#&gt; f4info: </span></span>
<span><span class="co">#&gt; f4rank: 1 dof:      1 chisq:     2.144 tail:          0.143115702 dofdiff:      3 chisqdiff:    -2.144 taildiff:                    1</span></span>
<span><span class="co">#&gt; B:</span></span>
<span><span class="co">#&gt;           scale     1.000 </span></span>
<span><span class="co">#&gt;        Denisova     0.475 </span></span>
<span><span class="co">#&gt;           Altai     1.332 </span></span>
<span><span class="co">#&gt; A:</span></span>
<span><span class="co">#&gt;           scale    27.595 </span></span>
<span><span class="co">#&gt;         Vindija     1.414 </span></span>
<span><span class="co">#&gt;          Yoruba    -0.030 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; full rank</span></span>
<span><span class="co">#&gt; f4info: </span></span>
<span><span class="co">#&gt; f4rank: 2 dof:      0 chisq:     0.000 tail:                    1 dofdiff:      1 chisqdiff:     2.144 taildiff:          0.143115702</span></span>
<span><span class="co">#&gt; B:</span></span>
<span><span class="co">#&gt;           scale    19.566   749.540 </span></span>
<span><span class="co">#&gt;        Denisova     0.470    -0.811 </span></span>
<span><span class="co">#&gt;           Altai     1.334    -1.158 </span></span>
<span><span class="co">#&gt; A:</span></span>
<span><span class="co">#&gt;           scale     1.414     1.414 </span></span>
<span><span class="co">#&gt;         Vindija     1.414     0.000 </span></span>
<span><span class="co">#&gt;          Yoruba     0.000     1.414 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; best coefficients:     0.021     0.979 </span></span>
<span><span class="co">#&gt; Jackknife mean:      0.020626834     0.979373166 </span></span>
<span><span class="co">#&gt;       std. errors:     0.006     0.006 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; error covariance (* 1,000,000)</span></span>
<span><span class="co">#&gt;         34        -34 </span></span>
<span><span class="co">#&gt;        -34         34 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; summ: Han    2      0.143116     0.021     0.979         34        -34         34 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     fixed pat  wt  dof     chisq       tail prob</span></span>
<span><span class="co">#&gt;            00  0     1     2.144        0.143116     0.021     0.979 </span></span>
<span><span class="co">#&gt;            01  1     2 14965.791               0     1.000     0.000 </span></span>
<span><span class="co">#&gt;            10  1     2    14.454     0.000726865     0.000     1.000 </span></span>
<span><span class="co">#&gt; best pat:           00         0.143116              -  -</span></span>
<span><span class="co">#&gt; best pat:           10      0.000726865  chi(nested):    12.309 p-value for nested model:     0.000450682</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; coeffs:     0.021     0.979 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ## dscore:: f_4(Base, Fit, Rbase, right2)</span></span>
<span><span class="co">#&gt; ## genstat:: f_4(Base, Fit, right1, right2)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; details:              Vindija             Denisova     0.024008   40.874863</span></span>
<span><span class="co">#&gt; details:               Yoruba             Denisova    -0.001082   -2.595761</span></span>
<span><span class="co">#&gt; dscore:             Denisova f4:    -0.000565 Z:    -1.366415</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; details:              Vindija                Altai     0.068176  121.272343</span></span>
<span><span class="co">#&gt; details:               Yoruba                Altai    -0.001546   -3.724300</span></span>
<span><span class="co">#&gt; dscore:                Altai f4:    -0.000109 Z:    -0.264364</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; gendstat:                Chimp             Denisova    -1.366</span></span>
<span><span class="co">#&gt; gendstat:                Chimp                Altai    -0.264</span></span>
<span><span class="co">#&gt; gendstat:             Denisova                Altai     1.131</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ##end of qpAdm:        1.674 seconds cpu        0.000 Mbytes in use</span></span></code></pre></div>
<p>Finally, we might want to keep some of the log files around for
future reference, further debugging or sharing with others. For these
purposes, we can run the <code>loginfo</code> function with the argument
<code>save = TRUE</code>, potentially also specifying the output
directory (<code>dir = "."</code> by default) or a prefix of the output
file(s) (default prefix is simply the name of the admixr command which
generated the log file).</p>
<p>For example, the following will save the result of our qpAdm analysis
of Neandertal ancestry to a file with a prefix
“qpAdm_Neandertal_ancestry”, but just the one for the Sardinian
individual:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/loginfo.html">loginfo</a></span><span class="op">(</span><span class="va">qpadm_res</span>, target <span class="op">=</span> <span class="st">"Sardinian"</span>, save <span class="op">=</span> <span class="cn">TRUE</span>, prefix <span class="op">=</span> <span class="st">"qpAdm_Neandertal_ancestry"</span><span class="op">)</span></span></code></pre></div>
<p>This will create a file in the current directory named
<code>qpAdm_Neandertal_ancestry_Sardinian.txt</code>. Note that when you
call <code><a href="../reference/loginfo.html">loginfo()</a></code> on a qpAdm run, the name of the target is
always added to the end of the output log file.</p>
<hr>
<p><a rel="license" href="https://creativecommons.org/licenses/by/4.0/" class="external-link"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png"></a></p>
<p>This work is licensed under a
<a rel="license" href="https://creativecommons.org/licenses/by/4.0/" class="external-link">Creative
Commons Attribution 4.0 International License</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
