<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>
<code>admixr</code> - overview â€¢ admixr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="&lt;code&gt;admixr&lt;/code&gt; - overview">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">admixr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/overview.html">`admixr` - overview</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bodkan/admixr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>
<code>admixr</code> - overview</h1>
                        <h4 class="author">Martin Petr</h4>
            
            <h4 class="date">2018-08-24</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/admixr/blob/master/vignettes/overview.Rmd"><code>vignettes/overview.Rmd</code></a></small>
      <div class="hidden name"><code>overview.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p><a href="http://www.genetics.org/content/192/3/1065">ADMIXTOOLS</a> is a widely used software package for calculating statistics and testing hypotheses about population admixture. However, although powerful and comprehensive, it is not exactly user-friendly.</p>
<p>A typical ADMIXTOOLS workflow usually involves a combination of <code>sed</code>/<code>awk</code>/shell scripting and manual editing to create different configuration files. These are then passed as command-line arguments to one of ADMIXTOOLSâ€™ commands, controlling how to run a particular analysis. The result of each analysis is then redirected to another file, and the user needs to extract values of interest from this file (which is full of redundant information), most likely using more shell scripting or (worse) by manual copy-pasting. The results are then analyse in R, Excel or another program.</p>
<p>This workflow very cumbersome, especially if one wants to explore many hypotheses involving different combinations of populations. Most importantly, however, it makes it difficult to follow good practices of reproducible science.</p>
<p>This R package makes it possible to perform all stages of ADMIXTOOLS analysis entirely from within R. It provides a set of convenient functions that completely abstract away the need for â€œlow levelâ€ configuration of individual ADMIXTOOLS programs, allowing the user to focus on the analysis itself.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>To install <code>admixr</code> from Github you need to install the package <code>devtools</code> first. You can simply run:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">install.packages</span>(<span class="st">"devtools"</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"bodkan/admixr"</span>)</a></code></pre></div>
<p>Furthermore, if you want to follow the examples in this vignette, you will need the tidyverse meta-package, which you can install with:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">install.packages</span>(<span class="st">"tidyverse"</span>)</a></code></pre></div>
<p>When everything is ready, you can load both packages:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">library</span>(admixr)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="co">#&gt; â”€â”€ Attaching packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse 1.2.1 â”€â”€</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co">#&gt; âœ” ggplot2 3.0.0     âœ” purrr   0.2.5</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="co">#&gt; âœ” tibble  1.4.2     âœ” dplyr   0.7.6</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="co">#&gt; âœ” tidyr   0.8.1     âœ” stringr 1.3.1</span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="co">#&gt; âœ” readr   1.1.1     âœ” forcats 0.3.0</span></a>
<a class="sourceLine" id="cb3-8" data-line-number="8"><span class="co">#&gt; â”€â”€ Conflicts â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse_conflicts() â”€â”€</span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="co">#&gt; âœ– dplyr::filter() masks stats::filter()</span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="co">#&gt; âœ– dplyr::lag()    masks stats::lag()</span></a></code></pre></div>
</div>
<div id="a-boring-technical-note-about-eigenstrat-files-and-admixtools" class="section level2">
<h2 class="hasAnchor">
<a href="#a-boring-technical-note-about-eigenstrat-files-and-admixtools" class="anchor"></a>A boring technical note about EIGENSTRAT files and ADMIXTOOLS</h2>
<p>If you have an EIGENSTRAT â€œtripletâ€ of files ready and want to know how to calculate different statistics, feel free to skip this section.</p>
<div id="eigenstrat-file-format" class="section level3">
<h3 class="hasAnchor">
<a href="#eigenstrat-file-format" class="anchor"></a>EIGENSTRAT file format</h3>
<p>ADMIXTOOLS softwre uses a peculiar set of genetic file formats, which feel strange if you are used to working with <a href="http://samtools.github.io/hts-specs/VCFv4.3.pdf">VCF files</a>. However, the basic idea remains the same - we want to store and access SNP data (REF/ALT alleles) of a set of individuals at a defined set of genomic positions.</p>
<p>EIGENSTRAT datasets always contain three kinds of files:</p>
<ul>
<li>
<code>ind</code> file - specifies name, sex and population assignment of each sample</li>
<li>
<code>snp</code> file - specifies positions of SNPs, REF/ALT alleles etc.</li>
<li>
<code>geno</code> file - specifies SNP data in a dense string-based format:
<ul>
<li>0: homozygous ALT</li>
<li>1: heterozygote</li>
<li>2: homozygous REF</li>
<li>9: missing site</li>
</ul>
</li>
</ul>
<p>As you can see, VCF files are essentially a combination of all three files in a single file. Luckily for us, all three EIGENSTRAT files usually share a common path and prefix (at least you should try to make it so whenever you work with them). This allows us to work with just the prefix, instead of worrying about individual files.</p>
<p>As such, all main <code>admixr</code> functions accept an argument called <code>prefix</code>, which specifies the path to and prefix of all three EIGENSTRAT files (you can still work with individual files if you need to, using <code>ind</code>, <code>snp</code> and <code>geno</code> arguments of each <code>admixr</code> function, but try to avoid that).</p>
<p>Here is a prefix of a small testing SNP dataset thatâ€™s distributed with <code>admixr</code>. We will be working with this data in the rest of this vignette.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">(eigen &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">"admixr"</span>, <span class="st">"extdata"</span>), <span class="st">"snps"</span>))</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="co">#&gt; [1] "/Users/martin_petr/local/R_LIBS/admixr/extdata/snps"</span></a></code></pre></div>
<p>We can verify that there are three files with this prefix:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">dir</span>(<span class="dt">path =</span> <span class="kw">dirname</span>(eigen))</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="co">#&gt; [1] "snps.geno" "snps.ind"  "snps.snp"</span></a></code></pre></div>
<p>Letâ€™s look at their contents. Here are three lines from each of them:</p>
<p><strong><code>ind</code> file</strong></p>
<pre><code>#&gt; Chimp    U   Chimp
#&gt; Mbuti    U   Mbuti
#&gt; Yoruba   U   Yoruba</code></pre>
<p><strong><code>snp</code> file</strong></p>
<pre><code>#&gt; 1_777122 1   0.007771    777122  A   T
#&gt; 1_832756 1   0.008328    832756  T   G
#&gt; 1_834832 1   0.008348    834832  G   C</code></pre>
<p><strong><code>geno</code> file</strong></p>
<pre><code>#&gt; 9220000000000
#&gt; 9022021012000
#&gt; 9022021012000</code></pre>
</div>
</div>
<div id="philosophy-of-admixr" class="section level2">
<h2 class="hasAnchor">
<a href="#philosophy-of-admixr" class="anchor"></a>Philosophy of <code>admixr</code>
</h2>
<p>The goal of this package is to make it easier to run ADMIXTOOLS analyses, without having to worry about (in ADMIXTOOLS jargon) par/pop/log/left/right files and other low level detail.</p>
<p>The only interface between you and ADMIXTOOLS are the following R functions:</p>
<ul>
<li><code><a href="../reference/d.html">d()</a></code></li>
<li><code><a href="../reference/f4.html">f4()</a></code></li>
<li><code><a href="../reference/f4ratio.html">f4ratio()</a></code></li>
<li><code><a href="../reference/f3.html">f3()</a></code></li>
<li><code><a href="../reference/qpAdm.html">qpAdm()</a></code></li>
</ul>
<p>Anything that would often take <a href="https://gaworkshop.readthedocs.io/en/latest/contents/06_f3/f3.html">dozens of lines of shell scripts</a> can be accomplished by running a single line of R code.</p>
<p>The only thing that is needed is:</p>
<ul>
<li>triplet of files in an EIGENSTRAT format</li>
<li>working ADMIXTOOLS installation</li>
<li>R :)</li>
</ul>
</div>
<div id="d-statistic" class="section level2">
<h2 class="hasAnchor">
<a href="#d-statistic" class="anchor"></a><span class="math inline">\(D\)</span> statistic</h2>
<p>Next question we could ask is - how closely related are different African populations to non-Africans today? One way to look at this is using the following D statistic: <span class="math inline">\(D(\textrm{non-African}, \textrm{Denisova}, \textrm{African Y}, \textrm{Chimp})\)</span>. This statistic is simply based on counting sites that match an African <span class="math inline">\(Y\)</span>, and is informative about the â€œbranchingâ€ order of non-Africans from non-Africans over time. The more positive this <span class="math inline">\(D\)</span> statistic is, the more recently did non-Africans split from an African population $Y4.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="co"># place sample names into variables for more readable code</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2">nonafr &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"French"</span>, <span class="st">"Sardinian"</span>, <span class="st">"Han"</span>, <span class="st">"Papuan"</span>)</a>
<a class="sourceLine" id="cb9-3" data-line-number="3">afr &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"Khomani_San"</span>, <span class="st">"Mbuti"</span>, <span class="st">"Yoruba"</span>, <span class="st">"Dinka"</span>)</a></code></pre></div>
<p>We can calculate <span class="math inline">\(D\)</span> statistic using <code>admixr</code>â€™s function <code>d</code>:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">df &lt;-<span class="st"> </span><span class="kw"><a href="../reference/d.html">d</a></span>(<span class="dt">W =</span> nonafr, <span class="dt">X =</span> <span class="st">"Denisova"</span>, <span class="dt">Y =</span>  afr, <span class="dt">Z =</span> <span class="st">"Chimp"</span>, <span class="dt">prefix =</span> eigen)</a></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw">head</span>(df)</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">W</th>
<th align="left">X</th>
<th align="left">Y</th>
<th align="left">Z</th>
<th align="right">D</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
<th align="right">BABA</th>
<th align="right">ABBA</th>
<th align="right">nsnps</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">French</td>
<td align="left">Denisova</td>
<td align="left">Khomani_San</td>
<td align="left">Chimp</td>
<td align="right">0.3922</td>
<td align="right">0.007191</td>
<td align="right">54.543</td>
<td align="right">13710</td>
<td align="right">5985</td>
<td align="right">194959</td>
</tr>
<tr class="even">
<td align="left">French</td>
<td align="left">Denisova</td>
<td align="left">Mbuti</td>
<td align="left">Chimp</td>
<td align="right">0.4255</td>
<td align="right">0.006923</td>
<td align="right">61.465</td>
<td align="right">14683</td>
<td align="right">5918</td>
<td align="right">194981</td>
</tr>
<tr class="odd">
<td align="left">French</td>
<td align="left">Denisova</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.4793</td>
<td align="right">0.006408</td>
<td align="right">74.785</td>
<td align="right">16533</td>
<td align="right">5820</td>
<td align="right">194985</td>
</tr>
<tr class="even">
<td align="left">French</td>
<td align="left">Denisova</td>
<td align="left">Dinka</td>
<td align="left">Chimp</td>
<td align="right">0.5158</td>
<td align="right">0.006256</td>
<td align="right">82.441</td>
<td align="right">17286</td>
<td align="right">5522</td>
<td align="right">195010</td>
</tr>
<tr class="odd">
<td align="left">Sardinian</td>
<td align="left">Denisova</td>
<td align="left">Khomani_San</td>
<td align="left">Chimp</td>
<td align="right">0.3915</td>
<td align="right">0.007093</td>
<td align="right">55.196</td>
<td align="right">13924</td>
<td align="right">6089</td>
<td align="right">194883</td>
</tr>
<tr class="even">
<td align="left">Sardinian</td>
<td align="left">Denisova</td>
<td align="left">Mbuti</td>
<td align="left">Chimp</td>
<td align="right">0.4254</td>
<td align="right">0.006664</td>
<td align="right">63.838</td>
<td align="right">14918</td>
<td align="right">6013</td>
<td align="right">194908</td>
</tr>
</tbody>
</table>
<p>We can see that the resulting <code>data.frame</code> object contains all the input information, but also additional statistics:</p>
<ul>
<li>
<code>D</code> - <span class="math inline">\(D\)</span> statistic value</li>
<li>
<code>stderr</code> - standard error of the <span class="math inline">\(D\)</span> statistic calculated with block jackknife</li>
<li>
<code>Zscore</code> - <span class="math inline">\(Z\)</span> significance value</li>
<li>
<code>BABA</code>/<code>ABBA</code> - counts of observed site patterns</li>
<li>
<code>nsnps</code> - number of SNPs used for the calculation in this row</li>
</ul>
<p>The format of output tables is very similar of other <code>admixr</code> functions, and they generally differ only in the name of the output column.</p>
<p>In general, tables are not a good representation of this kind of data, especially as the number of samples increases. This is what the results look like when plotted using the <code>ggplot2</code> package:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="kw">fct_reorder</span>(Y, D), D, <span class="dt">color =</span> W)) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb12-3" data-line-number="3"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">"African population Y"</span>)</a></code></pre></div>
<p><img src="overview_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
<p>We can see that the closest African population to non-Africans today are East African <a href="https://en.wikipedia.org/wiki/Dinka">Dinka</a>, which is consistent with the hypothesis that the ancestors of all non-Africans today split of the East African lineage during the Out of Africa migration. Decreasing values of the <span class="math inline">\(D\)</span> statistic for Yoruba, Mbuti and San fits with what we know about the split times of African from <a href="http://science.sciencemag.org/content/early/2017/09/27/science.aao6266">other</a> <a href="https://www.cell.com/cell/abstract/S0092-8674(17)31008-5">studies</a>, with San being the most divergent modern human population, followed by Mbuti, Yoruba and Dinka.</p>
</div>
<div id="f_4-statistic" class="section level2">
<h2 class="hasAnchor">
<a href="#f_4-statistic" class="anchor"></a><span class="math inline">\(f_4\)</span> statistic</h2>
<p>An alternative way of addressing the previous question would be to use the <span class="math inline">\(f_4\)</span> statistic, which is very similar to <span class="math inline">\(D\)</span> statistic. In fact, if <span class="math inline">\(D\)</span> is calculated as</p>
<p><span class="math display">\[ D = \frac{\textrm{# BABA sites - # ABBA sites}}{\textrm{# BABA sites + # ABBA sites}},\]</span></p>
<p><span class="math inline">\(f_4\)</span> statistic is simply obtained by skipping the normalization step. Therefore,</p>
<p><span class="math display">\[ f_4 = \textrm{# BABA sites - # ABBA sites}\]</span></p>
<p>Sometime one is more convenient than other though, so <code>admixr</code> provides a separate function for calculating <span class="math inline">\(f_4\)</span> statistics called <code>f4</code>.</p>
<p>To repeat the previous analysis, we can run:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">df &lt;-<span class="st"> </span><span class="kw"><a href="../reference/f4.html">f4</a></span>(<span class="dt">W =</span> nonafr, <span class="dt">X =</span> <span class="st">"Denisova"</span>, <span class="dt">Y =</span>  afr, <span class="dt">Z =</span> <span class="st">"Chimp"</span>, <span class="dt">prefix =</span> eigen)</a></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw">head</span>(df)</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">W</th>
<th align="left">X</th>
<th align="left">Y</th>
<th align="left">Z</th>
<th align="right">f4</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
<th align="right">BABA</th>
<th align="right">ABBA</th>
<th align="right">nsnps</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">French</td>
<td align="left">Denisova</td>
<td align="left">Khomani_San</td>
<td align="left">Chimp</td>
<td align="right">0.039622</td>
<td align="right">0.000797</td>
<td align="right">49.704</td>
<td align="right">13710</td>
<td align="right">5985</td>
<td align="right">194959</td>
</tr>
<tr class="even">
<td align="left">French</td>
<td align="left">Denisova</td>
<td align="left">Mbuti</td>
<td align="left">Chimp</td>
<td align="right">0.044957</td>
<td align="right">0.000784</td>
<td align="right">57.329</td>
<td align="right">14683</td>
<td align="right">5918</td>
<td align="right">194981</td>
</tr>
<tr class="odd">
<td align="left">French</td>
<td align="left">Denisova</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.054941</td>
<td align="right">0.000851</td>
<td align="right">64.572</td>
<td align="right">16533</td>
<td align="right">5820</td>
<td align="right">194985</td>
</tr>
<tr class="even">
<td align="left">French</td>
<td align="left">Denisova</td>
<td align="left">Dinka</td>
<td align="left">Chimp</td>
<td align="right">0.060324</td>
<td align="right">0.000889</td>
<td align="right">67.865</td>
<td align="right">17286</td>
<td align="right">5522</td>
<td align="right">195010</td>
</tr>
<tr class="odd">
<td align="left">Sardinian</td>
<td align="left">Denisova</td>
<td align="left">Khomani_San</td>
<td align="left">Chimp</td>
<td align="right">0.040205</td>
<td align="right">0.000804</td>
<td align="right">50.007</td>
<td align="right">13924</td>
<td align="right">6089</td>
<td align="right">194883</td>
</tr>
<tr class="even">
<td align="left">Sardinian</td>
<td align="left">Denisova</td>
<td align="left">Mbuti</td>
<td align="left">Chimp</td>
<td align="right">0.045688</td>
<td align="right">0.000784</td>
<td align="right">58.260</td>
<td align="right">14918</td>
<td align="right">6013</td>
<td align="right">194908</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="kw">fct_reorder</span>(Y, f4), f4, <span class="dt">color =</span> W)) <span class="op">+</span></a>
<a class="sourceLine" id="cb15-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb15-3" data-line-number="3"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">"African population Y"</span>)</a></code></pre></div>
<p><img src="overview_files/figure-html/unnamed-chunk-17-1.png" width="672"></p>
<p>As we can see, we can make the same conclusion using <span class="math inline">\(f_4\)</span> statistic, as we made using <span class="math inline">\(D\)</span> statistic. The higher the <span class="math inline">\(f_4\)</span> value, the more drift the samples in <span class="math inline">\(W\)</span> and <span class="math inline">\(Y\)</span> shared. In other words, the more recently the split between the two happened. Again, even this very simple analysis reproduces the major characteristics of modern human demographic history inferred using much more sophisticated methods.</p>
</div>
<div id="f_4-ratio-statistic" class="section level2">
<h2 class="hasAnchor">
<a href="#f_4-ratio-statistic" class="anchor"></a><span class="math inline">\(f_4\)</span>-ratio statistic</h2>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">df &lt;-<span class="st"> </span><span class="kw"><a href="../reference/f4ratio.html">f4ratio</a></span>(</a>
<a class="sourceLine" id="cb16-2" data-line-number="2">  <span class="dt">X =</span> <span class="kw">c</span>(<span class="st">"French"</span>, <span class="st">"Sardinian"</span>, <span class="st">"Han"</span>, <span class="st">"Papuan"</span>, <span class="st">"Dinka"</span>, <span class="st">"Mbuti"</span>, <span class="st">"Khomani_San"</span>),</a>
<a class="sourceLine" id="cb16-3" data-line-number="3">  <span class="dt">A =</span> <span class="st">"Altai"</span>, <span class="dt">B =</span> <span class="st">"Vindija"</span>, <span class="dt">C =</span> <span class="st">"Yoruba"</span>, <span class="dt">O =</span> <span class="st">"Chimp"</span>,</a>
<a class="sourceLine" id="cb16-4" data-line-number="4">  <span class="dt">prefix =</span> eigen</a>
<a class="sourceLine" id="cb16-5" data-line-number="5">)</a></code></pre></div>
<p>This is what the final <code>data.frame</code> looks like:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">df</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">A</th>
<th align="left">B</th>
<th align="left">X</th>
<th align="left">C</th>
<th align="left">O</th>
<th align="right">alpha</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">French</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.017916</td>
<td align="right">0.007773</td>
<td align="right">2.305</td>
</tr>
<tr class="even">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Sardinian</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.016491</td>
<td align="right">0.007376</td>
<td align="right">2.236</td>
</tr>
<tr class="odd">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Han</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.028865</td>
<td align="right">0.007855</td>
<td align="right">3.675</td>
</tr>
<tr class="even">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Papuan</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.041625</td>
<td align="right">0.007808</td>
<td align="right">5.331</td>
</tr>
<tr class="odd">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Dinka</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">-0.004999</td>
<td align="right">0.007186</td>
<td align="right">-0.696</td>
</tr>
<tr class="even">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Mbuti</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.004688</td>
<td align="right">0.007357</td>
<td align="right">0.637</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="kw">fct_reorder</span>(X, alpha), alpha, <span class="dt">color =</span> <span class="kw">abs</span>(Zscore) <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb18-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb18-3" data-line-number="3"><span class="st">  </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> alpha <span class="op">-</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>stderr, <span class="dt">ymax =</span> alpha <span class="op">+</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>stderr)) <span class="op">+</span></a>
<a class="sourceLine" id="cb18-4" data-line-number="4"><span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="dv">0</span>, <span class="dt">linetype =</span> <span class="dv">2</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb18-5" data-line-number="5"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">y =</span> <span class="st">"Neandertal ancestry proportion"</span>, <span class="dt">x =</span> <span class="st">"present-day individual"</span>)</a></code></pre></div>
<p><img src="overview_files/figure-html/unnamed-chunk-21-1.png" width="672"></p>
<p>Despite the wide confidence intervals (caused by a very low number of SNPs in our testing dataset), we can notice several well-known features:</p>
<ul>
<li>No significant Neandertal ancestry in present-day Africans.</li>
<li>Between 2-3% Neandertal ancestry in present-day non-Africans.</li>
<li>A much higher proportion of Neandertal ancestry in people from Papua New Guinea.</li>
</ul>
<p>Again, as you can see, we got a nice â€œpublication-readyâ€ figure in just a few lines of R code! No shell, no <code>awk/grep</code> or manual editing and copy-pasting required. ğŸ‘</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#installation">Installation</a></li>
      <li><a href="#a-boring-technical-note-about-eigenstrat-files-and-admixtools">A boring technical note about EIGENSTRAT files and ADMIXTOOLS</a></li>
      <li><a href="#philosophy-of-admixr">Philosophy of <code>admixr</code></a></li>
      <li><a href="#d-statistic"><span class="math inline">\(D\)</span> statistic</a></li>
      <li><a href="#f_4-statistic"><span class="math inline">\(f_4\)</span> statistic</a></li>
      <li><a href="#f_4-ratio-statistic"><span class="math inline">\(f_4\)</span>-ratio statistic</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
