<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Fitting qpAdm models with a 'rotation' strategy • admixr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Fitting qpAdm models with a 'rotation' strategy">
<meta property="og:description" content="admixr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">admixr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/qpAdm.html">Fitting qpAdm models with a 'rotation' strategy</a>
    </li>
    <li>
      <a href="../articles/tutorial.html">Tutorial and basic overview of the admixr R package</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bodkan/admixr/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="qpAdm_files/header-attrs-2.2/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Fitting qpAdm models with a ‘rotation’ strategy</h1>
                        <h4 class="author">Martin Petr</h4>
            
            <h4 class="date">2020-06-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/admixr/blob/master/vignettes/qpAdm.Rmd"><code>vignettes/qpAdm.Rmd</code></a></small>
      <div class="hidden name"><code>qpAdm.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>qpAdm model fitting is a complex topic. Navigating it successfuly requires solid knowledge of the <span class="math inline">\(f\)</span>-statistics theory first introduced by Nick Patterson and colleagues <a href="https://www.genetics.org/content/192/3/1065">in 2012</a>. As part of our <a href="tutorial.html#qpwave-and-qpadm-1">tutorial</a>, we have looked at a very basic overview of the qpAdm-related functionality implemented in <em>admixr</em> and also talked about some important resources for learning more the method pioneered by Iosif Lazaridis in <a href="https://www.nature.com/articles/nature14317">2015</a>.</p>
<p>Recently, Harney <em>et al.</em> published an exciting new preprint called <a href="https://www.biorxiv.org/content/10.1101/2020.04.09.032664v1">“Assessing the Performance of qpAdm: A Statistical Tool for Studying Population Admixture”</a>. Before we go any further, I encourage everyone to read it and the superb tutorial/guide available as its <a href="https://www.biorxiv.org/content/10.1101/2020.04.09.032664v1.supplementary-material">supplementary pdf</a> on bioRxiv. There really isn’t a better source of information on how to run and interpret qpAdm analyses.</p>
<p>Please, only attempt to run qpAdm if you have familiarized yoursef with all of the above-mentioned resources. I have had many people ask questions via email (not only about qpAdm but also other topics) to which the only sensible answer was - “you have to read the papers and understand the statistics first.” I know it’s frustrating but there really are no shortcuts here.</p>
</div>
<div id="qpadm-rotation" class="section level2">
<h2 class="hasAnchor">
<a href="#qpadm-rotation" class="anchor"></a><em>qpAdm</em> “rotation”</h2>
<p>If you have ever worked with <em>qpAdm</em>, you are well aware of the intricacies of finding the most suitable set of models that can explain the data. Among other things, we have to make a decision about the number of admixture sources and which populations are the most appropriate surrogates for those source populations (because only rarely we have sampled them directly). Furthermore, we need to carefully choose a number of so called ‘outgroup’ populations (also called ‘references’ or ‘right’ populations, depending on whom you talk to).</p>
<p>The preprint by <a href="https://www.biorxiv.org/content/10.1101/2020.04.09.032664v1">Harney <em>et al.</em></a> described an interesting idea to find a set of the most appropriate models (i.e. combinations of source and outgroup populations) which has been sucessfully used in the past. They call the method a “rotating population” strategy.</p>
<p>This approach starts by defining a set of “candidate” populations from which we iteratively sample a defined number of “sources” of ancestry for our “target” population of interest (most commonly two or three sources). After removing the sources from the candidate list, we then define all the remaining populations as “outgroups”. Finally, we iteratively fit qpAdm models for each combination of target, sources and outgroups, extracting <span class="math inline">\(p\)</span>-values and other statistics of interest. After finishing the exhaustive fitting of source-outgroup combinations, we examine all explored models, selecting those that seem most appropriate.</p>
</div>
<div id="implementation-in-admixr" class="section level2">
<h2 class="hasAnchor">
<a href="#implementation-in-admixr" class="anchor"></a>Implementation in <em>admixr</em>
</h2>
<p>In <em>admixr</em>, I have implemented a function <code><a href="../reference/qpAdm_rotation.html">qpAdm_rotation()</a></code> which does exactly what is described paragraph with one additional feature. Given the sensitivity of <em>qpAdm</em> to large numbers of potential outgroups (references), for each combination of sources and outgroups we also explore models for all possible <em>subsets</em> of outgroups. This is to find models which are as small as possible, possibly determining which outgroups are potentially redundant and not actually needed.</p>
<p>Let’s say that we have a target population <em>T</em> and a set of candidates for potential sources and outgroups <em>C</em> = {a, b, c, d, e, f}. Then, if we imagine an iteration of the rotation scheme in which we fixed sources <em>S</em> = {a, b}, we have remaining candidates for outgroups <em>C - S</em> = {c, d, e, f}. The basic implementation of the rotation procedure would simply take <em>C - S</em> as the full set of outgroups and fitted the following model:</p>
<ul>
<li>model #1: target <em>T</em>, sources <em>S</em> = {a, b} and outgroups = {c, d, e, f}</li>
</ul>
<p>However, in <em>admixr</em>, we would evaluate the following models in addition to the model #1:</p>
<ul>
<li>model #2: target <em>T</em>, sources <em>S</em> = {a, b} and outgroups = {c, d, e}</li>
<li>model #3: target <em>T</em>, sources <em>S</em> = {a, b} and outgroups = {c, d, f}</li>
<li>model #4: target <em>T</em>, sources <em>S</em> = {a, b} and outgroups = {c, e, f}</li>
<li>model #5: target <em>T</em>, sources <em>S</em> = {a, b} and outgroups = {d, e, f}.</li>
</ul>
<p>Therefore, our implementation in <code><a href="../reference/qpAdm_rotation.html">qpAdm_rotation()</a></code> explores all posible outgroup combinations, allowing us to look for the <em>smallest</em> model (in terms of outgroup size) that can explain our data.</p>
</div>
<div id="concrete-example" class="section level2">
<h2 class="hasAnchor">
<a href="#concrete-example" class="anchor"></a>Concrete example</h2>
<div id="performing-exhaustive-search-by-rotating-sourcesoutgroups" class="section level3">
<h3 class="hasAnchor">
<a href="#performing-exhaustive-search-by-rotating-sourcesoutgroups" class="anchor"></a>Performing exhaustive search by rotating sources/outgroups</h3>
<p>As an example, let’s revisit the problem of estimating the level of Neandertal ancestry in a French person from the main tutorial. We use this as an illustration because:</p>
<ol style="list-style-type: decimal">
<li>It’s the simplest possible analysis one could do with <em>qpAdm</em>.</li>
<li>It gives us a clear expectation of what the “truth” is.</li>
<li>It gives us a clear expectation of what models we should <em>definitely</em> reject.</li>
</ol>
<p>First, let’s download and install a development version of <em>admixr</em> to get access to the new features, and download a small example data set:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="co"># install `devtools` if not on already your system by running:</span>
<span class="co"># install.packages("devtools")</span>
<span class="co"># this will allow installing 'admixr' from Github</span>
<span class="co">#devtools::install_github("bodkan/admixr", ref = "qpadm-rotation", force = TRUE)</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">admixr</span>)

<span class="no">snps</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/eigenstrat.html">eigenstrat</a></span>(<span class="fu"><a href="../reference/download_data.html">download_data</a></span>())</pre></body></html></div>
<p>These are the individuals for which we have genotype data:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="../reference/read_ind.html">read_ind</a></span>(<span class="no">snps</span>)
<span class="co">#&gt; # A tibble: 12 x 3</span>
<span class="co">#&gt;    id          sex   label      </span>
<span class="co">#&gt;    &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;      </span>
<span class="co">#&gt;  1 Chimp       U     Chimp      </span>
<span class="co">#&gt;  2 Mbuti       U     Mbuti      </span>
<span class="co">#&gt;  3 Yoruba      U     Yoruba     </span>
<span class="co">#&gt;  4 Khomani_San U     Khomani_San</span>
<span class="co">#&gt;  5 Han         U     Han        </span>
<span class="co">#&gt;  6 Dinka       U     Dinka      </span>
<span class="co">#&gt;  7 Sardinian   U     Sardinian  </span>
<span class="co">#&gt;  8 Papuan      U     Papuan     </span>
<span class="co">#&gt;  9 French      U     French     </span>
<span class="co">#&gt; 10 Vindija     U     Vindija    </span>
<span class="co">#&gt; 11 Altai       U     Altai      </span>
<span class="co">#&gt; 12 Denisova    U     Denisova</span></pre></body></html></div>
<p>The <code><a href="../reference/qpAdm_rotation.html">qpAdm_rotation()</a></code> function is very simple. It accepts:</p>
<ul>
<li>a name of the target population,</li>
<li>a list of candidate populations,</li>
<li>a logical parameter <code>minimize</code>, determining whether to perform the “minimization” of the outgroup size described in the previous section,</li>
<li>the assumed number of sources of ancestry,</li>
<li>the number of CPU cores to use for analysis (be careful with this options as many ADMIXTOOLS analyses run in parallel can consume <em>a lot</em> of memory!),</li>
<li>parameter <code>fulloutput</code> specifying whether we want to have all the “ranks” and “subsets/patterns” statistics (see the main tutorial for more information) or if we just want the proportions of ancestry and significance values for individual models (this is the default, i.e. <code>fulloutput = FALSE</code>).</li>
</ul>
<p>So, let’s say we are interested in finding the proportions of archaic human ancestry in a French individual, and we also want to see what sorts of possible models we could find that match archaic introgression. We would run the following:</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">models</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/qpAdm_rotation.html">qpAdm_rotation</a></span>(
    <span class="kw">data</span> <span class="kw">=</span> <span class="no">snps</span>,
    <span class="kw">target</span> <span class="kw">=</span> <span class="st">"French"</span>,
    <span class="kw">candidates</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Dinka"</span>, <span class="st">"Mbuti"</span>, <span class="st">"Yoruba"</span>, <span class="st">"Vindija"</span>, <span class="st">"Altai"</span>, <span class="st">"Denisova"</span>, <span class="st">"Chimp"</span>),
    <span class="kw">minimize</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
    <span class="kw">nsources</span> <span class="kw">=</span> <span class="fl">2</span>,
    <span class="kw">ncores</span> <span class="kw">=</span> <span class="fl">2</span>,
    <span class="kw">fulloutput</span> <span class="kw">=</span> <span class="fl">TRUE</span>
)</pre></body></html></div>
<p>Here is what the full output looks like:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">models</span>
<span class="co">#&gt; $proportions</span>
<span class="co">#&gt; # A tibble: 336 x 13</span>
<span class="co">#&gt;    model target source1 source2 outgroups noutgroups  pvalue    prop1   prop2</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;          &lt;int&gt;   &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt;  1 m1    French Dinka   Mbuti   Yoruba &amp;…          3 3.37e-2    0.757   0.243</span>
<span class="co">#&gt;  2 m2    French Dinka   Mbuti   Yoruba &amp;…          3 1.38e-2    0.774   0.226</span>
<span class="co">#&gt;  3 m3    French Dinka   Mbuti   Yoruba &amp;…          3 1.14e-6    0.89    0.11 </span>
<span class="co">#&gt;  4 m4    French Dinka   Mbuti   Yoruba &amp;…          3 6.85e-2    0.781   0.219</span>
<span class="co">#&gt;  5 m5    French Dinka   Mbuti   Yoruba &amp;…          3 1.49e-5    0.885   0.115</span>
<span class="co">#&gt;  6 m6    French Dinka   Mbuti   Yoruba &amp;…          3 9.32e-3    0.887   0.113</span>
<span class="co">#&gt;  7 m7    French Dinka   Mbuti   Vindija …          3 6.70e-2   -8.51    9.51 </span>
<span class="co">#&gt;  8 m8    French Dinka   Mbuti   Vindija …          3 2.71e-1   25.3   -24.3  </span>
<span class="co">#&gt;  9 m9    French Dinka   Mbuti   Vindija …          3 1.61e-1 -187.    188.   </span>
<span class="co">#&gt; 10 m10   French Dinka   Mbuti   Altai &amp; …          3 8.10e-2  -30.7    31.7  </span>
<span class="co">#&gt; # … with 326 more rows, and 4 more variables: stderr1 &lt;dbl&gt;, stderr2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   nsnps_used &lt;dbl&gt;, nsnps_target &lt;dbl&gt;</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ranks</span>
<span class="co">#&gt; # A tibble: 672 x 9</span>
<span class="co">#&gt;    model target  rank    df chisq       tail dfdiff chisqdiff   taildiff</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;      &lt;dbl&gt;  &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 m1    French     1     1  4.51 0.0337          3     -4.51 1         </span>
<span class="co">#&gt;  2 m1    French     2     0  0    1               1      4.51 0.0337    </span>
<span class="co">#&gt;  3 m10   French     1     1  6.06 0.0138          3     -6.06 1         </span>
<span class="co">#&gt;  4 m10   French     2     0  0    1               1      6.06 0.0138    </span>
<span class="co">#&gt;  5 m100  French     1     1 23.7  0.00000114      3    -23.7  1         </span>
<span class="co">#&gt;  6 m100  French     2     0  0    1               1     23.7  0.00000114</span>
<span class="co">#&gt;  7 m101  French     1     1  3.32 0.0685          3     -3.32 1         </span>
<span class="co">#&gt;  8 m101  French     2     0  0    1               1      3.32 0.0685    </span>
<span class="co">#&gt;  9 m102  French     1     1 18.8  0.0000149       3    -18.8  1         </span>
<span class="co">#&gt; 10 m102  French     2     0  0    1               1     18.8  0.0000149 </span>
<span class="co">#&gt; # … with 662 more rows</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $subsets</span>
<span class="co">#&gt; # A tibble: 1,008 x 12</span>
<span class="co">#&gt;    model target source1 source2 pattern    wt   dof  chisq    tail prop1 prop2</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt;  1 m1    French Dinka   Mbuti   00          0     1   4.51 3.37e-2 0.757 0.243</span>
<span class="co">#&gt;  2 m1    French Dinka   Mbuti   01          1     2  23.6  7.35e-6 1     0    </span>
<span class="co">#&gt;  3 m1    French Dinka   Mbuti   10          1     2 160.   0.      0     1    </span>
<span class="co">#&gt;  4 m10   French Dinka   Mbuti   00          0     1   6.06 1.38e-2 0.774 0.226</span>
<span class="co">#&gt;  5 m10   French Dinka   Mbuti   01          1     2  23.3  8.91e-6 1     0    </span>
<span class="co">#&gt;  6 m10   French Dinka   Mbuti   10          1     2 162.   0.      0     1    </span>
<span class="co">#&gt;  7 m100  French Dinka   Mbuti   00          0     1  23.7  1.14e-6 0.89  0.11 </span>
<span class="co">#&gt;  8 m100  French Dinka   Mbuti   01          1     2  28.8  5.66e-7 1     0    </span>
<span class="co">#&gt;  9 m100  French Dinka   Mbuti   10          1     2 285.   0.      0     1    </span>
<span class="co">#&gt; 10 m101  French Dinka   Mbuti   00          0     1   3.32 6.85e-2 0.781 0.219</span>
<span class="co">#&gt; # … with 998 more rows, and 1 more variable: comment &lt;chr&gt;</span></pre></body></html></div>
<p>We can see a list with three components, as we would expect from any other <code><a href="../reference/qpAdm.html">qpAdm()</a></code> run (again, see the manual page and the tutorial for description of all three elements and their meaning). The first column of each component is always named <code>model</code> - this contains a short identifier of each individual “rotation” run (i.e., a combination target &amp; sources &amp; outgroups). It’s values don’t have any particular meaning - the order is completely arbitrary!, This variable is useful for later filtering and examination of individual models in detail.</p>
<p>Let’s ignore the <code>$ranks</code> and <code>$subsets</code> elements for now. We will focus only on the first element, <code>$proportions</code> which contains the main <em>qpAdm</em> summary.</p>
</div>
<div id="examining-and-filtering-fitted-models" class="section level3">
<h3 class="hasAnchor">
<a href="#examining-and-filtering-fitted-models" class="anchor"></a>Examining and filtering fitted models</h3>
<p>The <code>$proportions</code> table shown above contains information about <em>all</em> models, regardless of their plausibility. We can see that by examining the distributions of p-values (column <code>pvalue</code>) and admixture proportions (columns <code>prop1</code> and <code>prop2</code>) of each evaluated model in the figure below.</p>
<p>Notice two things (each dot represents one examined <em>qpAdm</em> model):</p>
<ul>
<li>Many models have inferred admixture proportions <em>way</em> outside the [0, 1] interval - those are clearly nonsensical.</li>
<li>Many models have very low p-values - this means these are incompatible with the data and can be rejected.</li>
</ul>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidyverse</span>)

<span class="fu">select</span>(<span class="no">models</span>$<span class="no">proportions</span>, <span class="no">model</span>, <span class="no">pvalue</span>, <span class="no">prop1</span>, <span class="no">prop2</span>) <span class="kw">%&gt;%</span>
    <span class="fu">gather</span>(<span class="no">parameter</span>, <span class="no">value</span>, -<span class="no">model</span>) <span class="kw">%&gt;%</span>
    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="no">parameter</span>, <span class="no">value</span>)) +
    <span class="fu">geom_jitter</span>() +
    <span class="fu">facet_wrap</span>(~ <span class="no">parameter</span>, <span class="kw">scales</span> <span class="kw">=</span> <span class="st">"free"</span>)</pre></body></html></div>
<p><img src="qpAdm_files/figure-html/unnamed-chunk-5-1.png" width="576"></p>
<p>To make it easier to narrow down the list of all models, <em>admixr</em> package contains a function <code><a href="../reference/qpAdm_filter.html">qpAdm_filter()</a></code>. This function accepts the result of the <code><a href="../reference/qpAdm_rotation.html">qpAdm_rotation()</a></code> function (either the <code>fulloutput = TRUE</code> version or the simple data frame with admixture proportions, p-values etc. produced by using<code>fulloutput = FALSE</code>) and filters out models with any of the proportions outside of the [0, 1] range and with p-values lower than a specified cutoff (0.05 by default):</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="co"># filter out models which can clearly be rejected</span>
<span class="no">fits</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/qpAdm_filter.html">qpAdm_filter</a></span>(<span class="no">models</span>)</pre></body></html></div>
<p>We can verify that the filtering worked by visualizing the filtered set of models again. Note that the p-values are distributed across the range of “insigificance” (i.e., “non-rejection”) between [0.05, 1.0]. Furthermore - remember that we originally set out to find combinations of sources-outgroups that model archaic ancestry in a French individual? We can clearly see two tidy clusters of estimated ancestry proportions. One is very small (this corresponds to the Neandertal component in modern humans - we would expect about 2-3% based on many previous analyses) and one large (“modern human” component, non-Neandertal ancestry):</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu">select</span>(<span class="no">fits</span>$<span class="no">proportions</span>, <span class="no">model</span>, <span class="no">pvalue</span>, <span class="no">prop1</span>, <span class="no">prop2</span>) <span class="kw">%&gt;%</span>
    <span class="fu">gather</span>(<span class="no">parameter</span>, <span class="no">value</span>, -<span class="no">model</span>) <span class="kw">%&gt;%</span>
    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="no">parameter</span>, <span class="no">value</span>)) +
    <span class="fu">geom_jitter</span>() +
    <span class="fu">facet_wrap</span>(~ <span class="no">parameter</span>, <span class="kw">scales</span> <span class="kw">=</span> <span class="st">"free"</span>) +
    <span class="fu">coord_cartesian</span>(<span class="kw">y</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">1</span>))</pre></body></html></div>
<p><img src="qpAdm_files/figure-html/unnamed-chunk-7-1.png" width="576"></p>
<p>Let’s now focus only on the proportions table. We will also ignore a couple of columns for brevity. Note that we are now also completely ignoring p-values because we <em>cannot</em> used those for model selection - they are <em>not</em> statistically meaningful at this stage! Higher p-value <em>never</em> implies higher likelihood of the model. Finally, we order the models based on the size of the outgroup set (smaller models first):</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">props</span> <span class="kw">&lt;-</span> <span class="no">fits</span>$<span class="no">proportions</span> <span class="kw">%&gt;%</span>
    <span class="fu">arrange</span>(<span class="no">noutgroups</span>) <span class="kw">%&gt;%</span>
    <span class="fu">select</span>(-<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">target</span>, <span class="no">noutgroups</span>, <span class="no">stderr1</span>, <span class="no">stderr2</span>, <span class="no">nsnps_used</span>, <span class="no">nsnps_target</span>))

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">props</span>, <span class="kw">n</span> <span class="kw">=</span> <span class="fl">Inf</span>)
<span class="co">#&gt; # A tibble: 56 x 7</span>
<span class="co">#&gt;    model source1 source2  outgroups                           pvalue prop1 prop2</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;                                &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt;  1 m81   Dinka   Chimp    Mbuti &amp; Yoruba &amp; Vindija            0.963  0.937 0.063</span>
<span class="co">#&gt;  2 m82   Dinka   Chimp    Mbuti &amp; Yoruba &amp; Altai              0.959  0.942 0.058</span>
<span class="co">#&gt;  3 m56   Dinka   Altai    Yoruba &amp; Vindija &amp; Chimp            0.898  0.973 0.027</span>
<span class="co">#&gt;  4 m40   Dinka   Vindija  Yoruba &amp; Altai &amp; Chimp              0.875  0.975 0.025</span>
<span class="co">#&gt;  5 m52   Dinka   Altai    Mbuti &amp; Vindija &amp; Denisova          0.829  0.977 0.023</span>
<span class="co">#&gt;  6 m183  Yoruba  Vindija  Mbuti &amp; Altai &amp; Denisova            0.826  0.993 0.007</span>
<span class="co">#&gt;  7 m83   Dinka   Chimp    Mbuti &amp; Yoruba &amp; Denisova           0.712  0.96  0.04 </span>
<span class="co">#&gt;  8 m65   Dinka   Denisova Mbuti &amp; Yoruba &amp; Vindija            0.712  0.96  0.04 </span>
<span class="co">#&gt;  9 m66   Dinka   Denisova Mbuti &amp; Yoruba &amp; Altai              0.672  0.963 0.037</span>
<span class="co">#&gt; 10 m41   Dinka   Vindija  Yoruba &amp; Denisova &amp; Chimp           0.603  0.97  0.03 </span>
<span class="co">#&gt; 11 m57   Dinka   Altai    Yoruba &amp; Denisova &amp; Chimp           0.603  0.97  0.03 </span>
<span class="co">#&gt; 12 m55   Dinka   Altai    Yoruba &amp; Vindija &amp; Denisova         0.599  0.974 0.026</span>
<span class="co">#&gt; 13 m199  Yoruba  Altai    Mbuti &amp; Vindija &amp; Denisova          0.562  0.991 0.009</span>
<span class="co">#&gt; 14 m36   Dinka   Vindija  Mbuti &amp; Altai &amp; Denisova            0.561  0.979 0.021</span>
<span class="co">#&gt; 15 m50   Dinka   Altai    Mbuti &amp; Yoruba &amp; Denisova           0.550  0.973 0.027</span>
<span class="co">#&gt; 16 m34   Dinka   Vindija  Mbuti &amp; Yoruba &amp; Denisova           0.548  0.973 0.027</span>
<span class="co">#&gt; 17 m49   Dinka   Altai    Mbuti &amp; Yoruba &amp; Vindija            0.527  0.975 0.025</span>
<span class="co">#&gt; 18 m33   Dinka   Vindija  Mbuti &amp; Yoruba &amp; Altai              0.504  0.977 0.023</span>
<span class="co">#&gt; 19 m51   Dinka   Altai    Mbuti &amp; Yoruba &amp; Chimp              0.460  0.981 0.019</span>
<span class="co">#&gt; 20 m35   Dinka   Vindija  Mbuti &amp; Yoruba &amp; Chimp              0.459  0.981 0.019</span>
<span class="co">#&gt; 21 m67   Dinka   Denisova Mbuti &amp; Yoruba &amp; Chimp              0.448  0.982 0.018</span>
<span class="co">#&gt; 22 m37   Dinka   Vindija  Mbuti &amp; Altai &amp; Chimp               0.400  0.977 0.023</span>
<span class="co">#&gt; 23 m58   Dinka   Altai    Vindija &amp; Denisova &amp; Chimp          0.367  0.972 0.028</span>
<span class="co">#&gt; 24 m39   Dinka   Vindija  Yoruba &amp; Altai &amp; Denisova           0.360  0.977 0.023</span>
<span class="co">#&gt; 25 m233  Yoruba  Chimp    Mbuti &amp; Altai &amp; Denisova            0.343  0.988 0.012</span>
<span class="co">#&gt; 26 m42   Dinka   Vindija  Altai &amp; Denisova &amp; Chimp            0.291  0.976 0.024</span>
<span class="co">#&gt; 27 m53   Dinka   Altai    Mbuti &amp; Vindija &amp; Chimp             0.290  0.974 0.026</span>
<span class="co">#&gt; 28 m38   Dinka   Vindija  Mbuti &amp; Denisova &amp; Chimp            0.198  0.975 0.025</span>
<span class="co">#&gt; 29 m54   Dinka   Altai    Mbuti &amp; Denisova &amp; Chimp            0.197  0.975 0.025</span>
<span class="co">#&gt; 30 m232  Yoruba  Chimp    Mbuti &amp; Vindija &amp; Denisova          0.123  0.986 0.014</span>
<span class="co">#&gt; 31 m202  Yoruba  Altai    Vindija &amp; Denisova &amp; Chimp          0.0983 0.974 0.026</span>
<span class="co">#&gt; 32 m68   Dinka   Denisova Mbuti &amp; Vindija &amp; Altai             0.0809 0.964 0.036</span>
<span class="co">#&gt; 33 m215  Yoruba  Denisova Mbuti &amp; Vindija &amp; Altai             0.0788 0.988 0.012</span>
<span class="co">#&gt; 34 m231  Yoruba  Chimp    Mbuti &amp; Vindija &amp; Altai             0.0751 0.98  0.02 </span>
<span class="co">#&gt; 35 m186  Yoruba  Vindija  Altai &amp; Denisova &amp; Chimp            0.0699 0.978 0.022</span>
<span class="co">#&gt; 36 m84   Dinka   Chimp    Mbuti &amp; Vindija &amp; Altai             0.0698 0.942 0.058</span>
<span class="co">#&gt; 37 m71   Dinka   Denisova Yoruba &amp; Vindija &amp; Altai            0.0687 0.96  0.04 </span>
<span class="co">#&gt; 38 m4    Dinka   Mbuti    Yoruba &amp; Altai &amp; Denisova           0.0685 0.781 0.219</span>
<span class="co">#&gt; 39 m87   Dinka   Chimp    Yoruba &amp; Vindija &amp; Altai            0.0608 0.941 0.059</span>
<span class="co">#&gt; 40 m74   Dinka   Denisova Vindija &amp; Altai &amp; Chimp             0.0507 0.922 0.078</span>
<span class="co">#&gt; 41 m63   Dinka   Altai    Yoruba &amp; Vindija &amp; Denisova &amp; Chimp 0.788  0.973 0.027</span>
<span class="co">#&gt; 42 m59   Dinka   Altai    Mbuti &amp; Yoruba &amp; Vindija &amp; Denisova 0.770  0.976 0.024</span>
<span class="co">#&gt; 43 m44   Dinka   Vindija  Mbuti &amp; Yoruba &amp; Altai &amp; Chimp      0.660  0.976 0.024</span>
<span class="co">#&gt; 44 m47   Dinka   Vindija  Yoruba &amp; Altai &amp; Denisova &amp; Chimp   0.612  0.976 0.024</span>
<span class="co">#&gt; 45 m43   Dinka   Vindija  Mbuti &amp; Yoruba &amp; Altai &amp; Denisova   0.607  0.978 0.022</span>
<span class="co">#&gt; 46 m60   Dinka   Altai    Mbuti &amp; Yoruba &amp; Vindija &amp; Chimp    0.559  0.974 0.026</span>
<span class="co">#&gt; 47 m62   Dinka   Altai    Mbuti &amp; Vindija &amp; Denisova &amp; Chimp  0.433  0.975 0.025</span>
<span class="co">#&gt; 48 m45   Dinka   Vindija  Mbuti &amp; Yoruba &amp; Denisova &amp; Chimp   0.426  0.974 0.026</span>
<span class="co">#&gt; 49 m61   Dinka   Altai    Mbuti &amp; Yoruba &amp; Denisova &amp; Chimp   0.425  0.974 0.026</span>
<span class="co">#&gt; 50 m46   Dinka   Vindija  Mbuti &amp; Altai &amp; Denisova &amp; Chimp    0.406  0.977 0.023</span>
<span class="co">#&gt; 51 m75   Dinka   Denisova Mbuti &amp; Yoruba &amp; Vindija &amp; Altai    0.196  0.962 0.038</span>
<span class="co">#&gt; 52 m91   Dinka   Chimp    Mbuti &amp; Yoruba &amp; Vindija &amp; Altai    0.194  0.941 0.059</span>
<span class="co">#&gt; 53 m239  Yoruba  Chimp    Mbuti &amp; Vindija &amp; Altai &amp; Denisova  0.101  0.988 0.012</span>
<span class="co">#&gt; 54 m93   Dinka   Chimp    Mbuti &amp; Yoruba &amp; Altai &amp; Denisova   0.0786 0.954 0.046</span>
<span class="co">#&gt; 55 m64   Dinka   Altai    Mbuti &amp; Yoruba &amp; Vindija &amp; Denisov… 0.629  0.974 0.026</span>
<span class="co">#&gt; 56 m48   Dinka   Vindija  Mbuti &amp; Yoruba &amp; Altai &amp; Denisova … 0.580  0.976 0.024</span></pre></body></html></div>
<p>Fun fact: notice in the table below that there are many models in which the chimpanzee was fitted as a source of ancestry! Interestingly, qpAdm used Chimp to infer archaic human ancestry. This is because you could think of Neandertal ancestry as an “ancestral component” of a modern human genome and the <em>qpAdm</em> rotation procedure therefore concludes that Chimpanzee is not be an unreasonable surrogate for a source population. Of course, we know there are better sources in our candidates set - we have the archaic humans!</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="no">props</span>, <span class="no">source1</span> <span class="kw">==</span> <span class="st">"Chimp"</span> <span class="kw">|</span> <span class="no">source2</span> <span class="kw">==</span> <span class="st">"Chimp"</span>)
<span class="co">#&gt; # A tibble: 11 x 7</span>
<span class="co">#&gt;    model source1 source2 outgroups                          pvalue prop1 prop2</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;                               &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt;  1 m81   Dinka   Chimp   Mbuti &amp; Yoruba &amp; Vindija           0.963  0.937 0.063</span>
<span class="co">#&gt;  2 m82   Dinka   Chimp   Mbuti &amp; Yoruba &amp; Altai             0.959  0.942 0.058</span>
<span class="co">#&gt;  3 m83   Dinka   Chimp   Mbuti &amp; Yoruba &amp; Denisova          0.712  0.96  0.04 </span>
<span class="co">#&gt;  4 m233  Yoruba  Chimp   Mbuti &amp; Altai &amp; Denisova           0.343  0.988 0.012</span>
<span class="co">#&gt;  5 m232  Yoruba  Chimp   Mbuti &amp; Vindija &amp; Denisova         0.123  0.986 0.014</span>
<span class="co">#&gt;  6 m231  Yoruba  Chimp   Mbuti &amp; Vindija &amp; Altai            0.0751 0.98  0.02 </span>
<span class="co">#&gt;  7 m84   Dinka   Chimp   Mbuti &amp; Vindija &amp; Altai            0.0698 0.942 0.058</span>
<span class="co">#&gt;  8 m87   Dinka   Chimp   Yoruba &amp; Vindija &amp; Altai           0.0608 0.941 0.059</span>
<span class="co">#&gt;  9 m91   Dinka   Chimp   Mbuti &amp; Yoruba &amp; Vindija &amp; Altai   0.194  0.941 0.059</span>
<span class="co">#&gt; 10 m239  Yoruba  Chimp   Mbuti &amp; Vindija &amp; Altai &amp; Denisova 0.101  0.988 0.012</span>
<span class="co">#&gt; 11 m93   Dinka   Chimp   Mbuti &amp; Yoruba &amp; Altai &amp; Denisova  0.0786 0.954 0.046</span></pre></body></html></div>
<p>Another interesting fact: notice that the rotating population procedure selected another plausible model characterizing the ancestry of the French individual. However, this of course doesn’t represent Neandertal introgression. What it might possibly represent is left as an exercise for the reader… :)</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="no">props</span>, <span class="no">prop1</span> <span class="kw">&lt;</span> <span class="fl">0.9</span>, <span class="no">prop2</span> <span class="kw">&lt;</span> <span class="fl">0.9</span>)
<span class="co">#&gt; # A tibble: 1 x 7</span>
<span class="co">#&gt;   model source1 source2 outgroups                 pvalue prop1 prop2</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;                      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 m4    Dinka   Mbuti   Yoruba &amp; Altai &amp; Denisova 0.0685 0.781 0.219</span></pre></body></html></div>
</div>
</div>
<div id="conclusions" class="section level2">
<h2 class="hasAnchor">
<a href="#conclusions" class="anchor"></a>Conclusions</h2>
<p>At this stage of analysis, you would have to decide which of the models produced by <code><a href="../reference/qpAdm_filter.html">qpAdm_filter()</a></code> that cannot be immediately rejected are more reasonable than others and why. Possibly based on both some prior knowledge and additional statistics (such as the details information available in the full log output information shown by <code><a href="../reference/loginfo.html">loginfo()</a></code>). You could say that the <em>qpAdm</em> methodology, while rooted in strong statistics, is from a certain point as much art as it is science. Interpreting the results and finding the most appropriate models can be quite a challenge.</p>
<p>Happy modeling and please, do <a href="https://github.com/bodkan/admixr/issues">let me know</a> if you discover bugs or missing features. My goal with this tool is to streamline <em>qpAdm</em> model fitting as much as possible and I can do it only with your input.</p>
</div>
<div id="final-remarks" class="section level2">
<h2 class="hasAnchor">
<a href="#final-remarks" class="anchor"></a>Final remarks</h2>
<ol style="list-style-type: decimal">
<li><p>As a reminder, keep in mind that <em>admixr</em> gives you tools for filtering SNPs and also grouping samples into populations on the fly! You can easily process and group samples before plugging them into <code><a href="../reference/qpAdm_rotation.html">qpAdm_rotation()</a></code>!</p></li>
<li><p>Also note that you can use the function <code><a href="../reference/loginfo.html">loginfo()</a></code> to examine the complete log output of any model by specifying the model identifier. This is helpful not only for debugging purposes but also for cases when you need a particular statistic in the full qpAdm log report which is not currently parsed by <em>admixr</em>:</p></li>
</ol>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="../reference/loginfo.html">loginfo</a></span>(<span class="no">fits</span>, <span class="st">"m40"</span>)
<span class="co">#&gt; Full output log of qpAdm rotation for model 'm40':</span>
<span class="co">#&gt; ===================================================</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; qpAdm: parameter file: /var/folders/t7/9gjtb6m92flbnp930618vt3r0000gn/T//Rtmpf6qd7v/qpAdm__1452457283.par</span>
<span class="co">#&gt; ### THE INPUT PARAMETERS</span>
<span class="co">#&gt; ##PARAMETER NAME: VALUE</span>
<span class="co">#&gt; genotypename: /var/folders/t7/9gjtb6m92flbnp930618vt3r0000gn/T//Rtmpf6qd7v/snps/snps.geno</span>
<span class="co">#&gt; snpname: /var/folders/t7/9gjtb6m92flbnp930618vt3r0000gn/T//Rtmpf6qd7v/snps/snps.snp</span>
<span class="co">#&gt; indivname: /var/folders/t7/9gjtb6m92flbnp930618vt3r0000gn/T//Rtmpf6qd7v/snps/snps.ind</span>
<span class="co">#&gt; popleft: /var/folders/t7/9gjtb6m92flbnp930618vt3r0000gn/T//Rtmpf6qd7v/qpAdm__1452457283.popleft</span>
<span class="co">#&gt; popright: /var/folders/t7/9gjtb6m92flbnp930618vt3r0000gn/T//Rtmpf6qd7v/qpAdm__1452457283.popright</span>
<span class="co">#&gt; allsnps: YES</span>
<span class="co">#&gt; summary: YES</span>
<span class="co">#&gt; details: YES</span>
<span class="co">#&gt; ## qpAdm version: 1000</span>
<span class="co">#&gt; seed: -1847170058</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; left pops:</span>
<span class="co">#&gt; French</span>
<span class="co">#&gt; Dinka</span>
<span class="co">#&gt; Vindija</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; right pops:</span>
<span class="co">#&gt; Yoruba</span>
<span class="co">#&gt; Altai</span>
<span class="co">#&gt; Chimp</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   0               French    1</span>
<span class="co">#&gt;   1                Dinka    1</span>
<span class="co">#&gt;   2              Vindija    1</span>
<span class="co">#&gt;   3               Yoruba    1</span>
<span class="co">#&gt;   4                Altai    1</span>
<span class="co">#&gt;   5                Chimp    1</span>
<span class="co">#&gt; jackknife block size:     0.050</span>
<span class="co">#&gt; snps: 500000  indivs: 6</span>
<span class="co">#&gt; number of blocks for block jackknife: 546</span>
<span class="co">#&gt; ## ncols: 500000</span>
<span class="co">#&gt; coverage:               French 499434</span>
<span class="co">#&gt; coverage:                Dinka 499362</span>
<span class="co">#&gt; coverage:              Vindija 497544</span>
<span class="co">#&gt; coverage:               Yoruba 499246</span>
<span class="co">#&gt; coverage:                Altai 497729</span>
<span class="co">#&gt; coverage:                Chimp 491273</span>
<span class="co">#&gt; dof (jackknife):   490.290</span>
<span class="co">#&gt; numsnps used: 500000</span>
<span class="co">#&gt; codimension 1</span>
<span class="co">#&gt; f4info: </span>
<span class="co">#&gt; f4rank: 1 dof:      1 chisq:     0.025 tail:          0.875038438 dofdiff:      3 chisqdiff:    -0.025 taildiff:                    1</span>
<span class="co">#&gt; B:</span>
<span class="co">#&gt;           scale     1.000 </span>
<span class="co">#&gt;           Altai     1.301 </span>
<span class="co">#&gt;           Chimp     0.554 </span>
<span class="co">#&gt; A:</span>
<span class="co">#&gt;           scale    15.252 </span>
<span class="co">#&gt;           Dinka    -0.036 </span>
<span class="co">#&gt;         Vindija     1.414 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; full rank</span>
<span class="co">#&gt; f4info: </span>
<span class="co">#&gt; f4rank: 2 dof:      0 chisq:     0.000 tail:                    1 dofdiff:      1 chisqdiff:     0.025 taildiff:          0.875038438</span>
<span class="co">#&gt; B:</span>
<span class="co">#&gt;           scale   421.013    10.787 </span>
<span class="co">#&gt;           Altai    -1.293     1.301 </span>
<span class="co">#&gt;           Chimp    -0.574     0.554 </span>
<span class="co">#&gt; A:</span>
<span class="co">#&gt;           scale     1.414     1.414 </span>
<span class="co">#&gt;           Dinka     1.414     0.000 </span>
<span class="co">#&gt;         Vindija     0.000     1.414 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; best coefficients:     0.975     0.025 </span>
<span class="co">#&gt; Jackknife mean:      0.975467965     0.024532035 </span>
<span class="co">#&gt;       std. errors:     0.005     0.005 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; error covariance (* 1,000,000)</span>
<span class="co">#&gt;         24        -24 </span>
<span class="co">#&gt;        -24         24 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; summ: French    2      0.875038     0.975     0.025         24        -24         24 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     fixed pat  wt  dof     chisq       tail prob</span>
<span class="co">#&gt;            00  0     1     0.025        0.875038     0.975     0.025 </span>
<span class="co">#&gt;            01  1     2    24.294     5.30367e-06     1.000     0.000 </span>
<span class="co">#&gt;            10  1     2 23181.059               0     0.000     1.000 </span>
<span class="co">#&gt; best pat:           00         0.875038              -  -</span>
<span class="co">#&gt; best pat:           01      5.30367e-06  chi(nested):    24.269 p-value for nested model:     8.37552e-07</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; coeffs:     0.975     0.025 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ## dscore:: f_4(Base, Fit, Rbase, right2)</span>
<span class="co">#&gt; ## genstat:: f_4(Base, Fit, right1, right2)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; details:                Dinka                Altai    -0.003070   -4.407893</span>
<span class="co">#&gt; details:              Vindija                Altai     0.120642  146.515551</span>
<span class="co">#&gt; dscore:                Altai f4:    -0.000037 Z:    -0.053286</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; details:                Dinka                Chimp    -0.001363   -2.294880</span>
<span class="co">#&gt; details:              Vindija                Chimp     0.051327   77.922651</span>
<span class="co">#&gt; dscore:                Chimp f4:    -0.000071 Z:    -0.121462</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; gendstat:               Yoruba                Altai    -0.053</span>
<span class="co">#&gt; gendstat:               Yoruba                Chimp    -0.121</span>
<span class="co">#&gt; gendstat:                Altai                Chimp    -0.086</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ##end of qpAdm:        3.948 seconds cpu        0.000 Mbytes in use</span></pre></body></html></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
